## The implementation of the prediction method presented in the study by Fourati et al. 2021
Fourati et al. (2021) [55] presented a RF model trained on a collection of transcriptomic profiles obtained from public databases covering vaccination against a wide range of pathogens (YFV, smallpox, influenza, pneumococcal, and meningococcal). In total the transcriptomics profiles were obtained from whole blood and peripheral blood mononuclear cells (PBMC) of 820 subjects before and after vaccination. The top 500 varying genes were used as input to the RF classifier and the class probabilities were utilized to predict the antibody response. It was possible to identify all 500 genes in the CMI-PB datasets. Gene symbols and Ensembl IDs for the top 500 genes are presented in Supplementary Figure S6. In the original study, the gene expression datasets were preprocessed with VST from DESeq, before being cross-study quantile normalized to match the Affymetrix distribution. Here, the CMI-PB gene expression count data were transformed with VST from DESeq2 and quantile normalized to match the distribution of the original study. The probability of being a high responder was utilized as the predictor for the response class in the original study, and hence there was an expected directionality of high responders obtaining higher probabilities from the RF classifier. 
