---
title: "Study 12 - Transcriptome profiling in blood before and after hepatitis B vaccination shows significant differences in gene expression between responders and non-responders (PMID: 30205979)"
author: "Mikkel Niklas Rasmussen"
date: "3/9/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Applying the 23 differentially expressed genes between the non-responder and responder at day 0 presented in study by Bartholomeus et al. (PMID: 30205979) on the CMI-PD dataset

```{r load libaries, echo=FALSE}
setwd(dirname(rstudioapi::getSourceEditorContext()$path))

suppressMessages(library('data.table'))
suppressMessages(library('dplyr'))
suppressMessages(library('plyr'))
suppressMessages(library('tidyr'))
suppressMessages(library('Biobase'))
suppressMessages(library('DESeq2'))
suppressMessages(library('pROC'))
suppressMessages(library("ggplot2"))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("biomaRt"))
suppressMessages(library('klaR'))
suppressMessages(library("tibble"))

# Load functions for evaluating gene signatures
source('../../R/functions/geneModelEvaluate.R')

```

#### Load the CMI-PB 2020 datasets

```{r load data}

DATA_DIR <- "../../Data/Pre_process/cmi-pb-multiomics/results/main/cmi_pb_datasets/processed/full"

# Info table (meta data)
meta.2020 <- fread(file.path(DATA_DIR, "meta.2020.pivoted.tsv.gz"))
meta.2021 <- fread(file.path(DATA_DIR, "meta.2021.pivoted.tsv.gz"))
specimen.table.2020 <- fread("../../Data/2021_cmipb_challenge/2020LD_specimen.csv")
specimen.table.2021 <- fread("../../Data/2021_cmipb_challenge/2021BD_specimen.csv")

# Antibody titer data
ab.titer.2020 <- fread("../../Data/2021_cmipb_challenge/2020LD_ab_titer.csv")
ab.titer.2021.day0 <- fread("../../Data/2021_cmipb_challenge/2021BD_ab_titer.csv")
ab.titer.2021.day14 <- fread("../../Data/2021_cmipb_challenge/2021_ab_titer_day14.csv")

# Gene expression data for day 0
ge.TPM.2020.day0 <- fread(file.path(DATA_DIR, "rnaseq_tpm.2020.day0.pivoted.tsv.gz"))
ge.RAW.2020.day0 <- fread(file.path(DATA_DIR, "rnaseq_raw.2020.day0.pivoted.tsv.gz"))

# Gene expression data for day 0 - 2021
ge.TPM.2021.day0 <- fread(file.path(DATA_DIR, "rnaseq_tpm.2021.day0.pivoted.tsv.gz"))
ge.RAW.2021.day0 <- fread(file.path(DATA_DIR, "rnaseq_raw.2021.day0.pivoted.tsv.gz"))

```

## Step 1: Preprocess CMI-PB 2020 antibody titer data

```{r}

#### Antibody titer data
# Normalized antibody titer data by baseline median
abt.2020.all <- ab.titer.2020 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14, 30)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data day 0 and 14
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14)

# Subset antibody titer
abt.2020.subset <- abt.2020.all[abt.2020.all$subject_id %in% ids, ]
abt.2020.day0 <- abt.2020.subset[abt.2020.subset$timepoint == 0, ]
abt.2020.day14 <- abt.2020.subset[abt.2020.subset$timepoint == 14, ]

```

## Step 2: Load differential genomic signature at day 0 

```{r}

# Load gene mapping 
study.gene.map <- read.delim("../hepb_genexp/data/genes.txt", header = FALSE,
                       col.names = c('gene_id', 'ncbi_entrez', 'uniprot_id',
                                     'gene_symbol', 'protein', 'synonyms',
                                     'description', 'ncbi_id', 'start', 'end'))

# Load differentially expressed genes
deg.genes <- read.table("../hepb_genexp/DEG_output/R-vs-NR-EXP0.csv", sep = ",",
                         header = TRUE) %>%
   dplyr::rename(id = X)

# Select the most significant DEGs as described in the original manuscript
significant.DEGs <- deg.genes[deg.genes$padj < 0.1, ] %>%
   drop_na()
sig.ids <- significant.DEGs$id

# Load gene signatures
gene.sig <- study.gene.map[study.gene.map$gene_id %in% significant.DEGs$id, ]$gene_symbol

```

## Step 3: Load TPM normalized gene expression data and normalized gene expression counts as in original study

```{r pre-process gene data}

# Tranpose the TPM normalized gene expression data
ge.TPM.2020.day0.T <- ge.TPM.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Tranpose the TPM normalized gene expression data
ge.RAW.2020.day0.T <- ge.RAW.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Normalization of gene expression as performed in original study
ge.RAW.2020.day0.T.long <- ge.RAW.2020.day0.T %>%
   as.data.frame() %>% 
   rownames_to_column("gene_id") %>% 
   pivot_longer(!gene_id, names_to = "subject_id", values_to = "count")

# Remove genes with sum count lower than 100
genecount_filter <- ge.RAW.2020.day0.T.long %>%
  group_by(gene_id) %>%
  dplyr::summarise(gene_count = sum(count))  %>%
  filter(gene_count > 100)

ge.RAW.2020.day0.T.long.filter <- ge.RAW.2020.day0.T.long %>%
  dplyr::filter(gene_id %in% genecount_filter$gene_id)

# Calculate medians count per subject
medcorr <- ge.RAW.2020.day0.T.long.filter  %>%
  group_by(subject_id) %>%
  mutate(count1 = ifelse(count == 0, NA, count)) %>%
  dplyr::summarise(n=n(), 
                   median = median(count), 
                   median_exlude_0 = median(count1, na.rm = T), 
                   sum=sum(count)) %>%
  dplyr::select(median_exlude_0, subject_id)

ge.RAW.2020.day0.T.long.median <- left_join(ge.RAW.2020.day0.T.long.filter, 
                                            medcorr, by="subject_id")

# Normalize the counts
ge.RAW.2020.day0.T.long.norm <- ge.RAW.2020.day0.T.long.median %>%
   dplyr::mutate(norm_count = ((count * 10000)/median_exlude_0)/10000)

# Convert back to normal matrix format
ge.RAW.2020.day0.T.norm <- ge.RAW.2020.day0.T.long.norm %>% 
   dplyr::select(gene_id, subject_id, norm_count) %>% 
   pivot_wider(names_from = subject_id, values_from = norm_count) %>% 
   column_to_rownames("gene_id")


```

## Step 4: Evaluate the predictive power of the differential genomic signature on day 14 antibody titers and fold-changes

```{r model evaluation}

# Isotype-antigens which are considered the most important
selected.iso.ag <- c("IgG-PT", "IgG-FHA",   "IgG-PRN", "IgG1-PT",
                     "IgG1-FHA", "IgG4-PT", "IgG4-FHA")


# Calculate the gene signature score with TPM normalized gene expression
df.gene.sig.2020.pred <- geneSigCalc(gene.ex.data=ge.TPM.2020.day0.T, 
                                     score='missing', 
                                     gene.sig=gene.sig,
                                     genes.up=NULL, 
                                     genes.down=NULL,
                                     post.ab.data=abt.2020.day14,
                                     pre.ab.data=abt.2020.day0,
                                     filter.ab.ag=selected.iso.ag,
                                     verbose=TRUE)

# Evaluate gene signature with TPM normalized gene expression
Bartholomeus.2018.gene.sig.2020.results <- evaluateModel(df.prediction=df.gene.sig.2020.pred, 
                                                         N.perm = 1000, direction = "missing",
                                                         add.to.title='- TPM normalized',
                                                         fn1='geneSig_vs_day14_heatmap_TPM',
                                                         fn2='geneSig_vs_day14_correlation_TPM',
                                                         fn3='geneSig_vs_FC_heatmap_TPM',
                                                         fn4='geneSig_vs_FC_correlation_TPM',
                                                         fn5='geneSig_vs_day14_AUC_TPM',
                                                         fn6='geneSig_vs_day14_boxplot_TPM',
                                                         fn7='geneSig_vs_FC_AUC_TPM',
                                                         fn8='geneSig_vs_FC_boxplot_TPM',
                                                         path = 'Results/2020/Gene_signature/')

save(Bartholomeus.2018.gene.sig.2020.results, 
     file = "Results/2020/Gene_signature/Bartholomeus.2018.gene.sig.2020.results.RData")

```

## Step 5: Evaluate the predictive power of the Naive Bayes classifier on day 14 antibody titers and fold-changes

```{r}

load("../hepb_genexp/predictor_scripts/classifier.RData")
model1 <- readRDS("../hepb_genexp/predictor_scripts/model1_without_granulocytes.rds")

# Calculate the class probability with the Naive Bayes classifier
df.NB.2020.pred <- mlModelPredict(gene.ex.data=ge.RAW.2020.day0.T.norm, 
                             study = 'Bartholomeus_2018', model=model1,
                             post.ab.data=abt.2020.day14,
                             pre.ab.data=abt.2020.day0,
                             filter.ab.ag=selected.iso.ag)

# Evaluate the class probability predicted by the Naive Bayes classifier
Bartholomeus.2018.NB.2020.results <- evaluateModel(df.prediction=df.NB.2020.pred, 
                                                   direction='missing',
                                                   N.perm = 1000,
                                                   score.label='Differential genomic signature score',
                                                   fn1='NBpred_vs_day14_heatmap_TPM',
                                                   fn2='NBpred_vs_day14_correlation_TPM',
                                                   fn3='NBpred_vs_FC_heatmap_TPM',
                                                   fn4='NBpred_vs_FC_correlation_TPM',
                                                   fn5='NBpred_vs_day14_AUC_TPM',
                                                   fn6='NBpred_vs_day14_boxplot_TPM',
                                                   fn7='NBpred_vs_FC_AUC_TPM',
                                                   fn8='NBpred_vs_FC_boxplot_TPM',
                                                   path = 'Results/2020/Naive_bayes/')
# Save results
save(Bartholomeus.2018.NB.2020.results, 
     file = "Results/2020/Naive_bayes/Bartholomeus.2018.NB.2020.results.RData")

```

## Step 6: Pre-process CMI-PB 2021 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
ab.titer.day0.2021.all <- ab.titer.2021.day0 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

ab.titer.subject.ids <- unique(ab.titer.day0.2021.all$subject_id)
abt.2021.day0 <- ab.titer.day0.2021.all[ab.titer.day0.2021.all$timepoint == 0, ]
abt.2021.day0 <- abt.2021.day0 %>% 
   filter(isotype_antigen %in% selected.iso.ag)

# Change format of the day 14 antibody titers
abt.2021.day14 <- ab.titer.2021.day14 %>% 
   separate(task, into=c("task", "antigen"), sep = "\\) ") %>% 
   dplyr::mutate(isotype_antigen = recode(antigen, 
                                          `Pertussis Toxin` = "IgG-PT", 
                                          `FHA` = 'IgG-FHA', 
                                          `Pertactin` =  'IgG-PRN',
                                          `IgG1 - Pertussis toxin` = "IgG1-PT",
                                          `IgG1 - FHA` = "IgG1-FHA",
                                          `IgG4 - Pertussis toxin` = "IgG4-PT",
                                          `IgG4 - FHA` = "IgG4-FHA"))

```

## Step 7: Transpose the CMI-PB 2021 TPM normalized expression data and normalized gene expression counts

```{r pre-process gene data}

# Subset gene expression data based on subjects with antibody titers
ge.TPM.2021.day0 <- ge.TPM.2021.day0[ge.TPM.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# Subset gene expression data based on subjects with antibody titers
ge.RAW.2021.day0 <- ge.RAW.2021.day0[ge.RAW.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# Tranpose the TPM normalized gene expression data
ge.TPM.2021.day0.T <- ge.TPM.2021.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Tranpose the TPM normalized gene expression data
ge.RAW.2021.day0.T <- ge.RAW.2021.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Normalization of gene expression from original study
ge.RAW.2021.day0.T.long <- ge.RAW.2021.day0.T %>%
   as.data.frame() %>% 
   rownames_to_column("gene_id") %>% 
   pivot_longer(!gene_id, names_to = "subject_id", values_to = "count")

# Remove genes with sum count lower than 100
genecount_filter <- ge.RAW.2021.day0.T.long %>%
  group_by(gene_id) %>%
  dplyr::summarise(gene_count = sum(count))  %>%
  filter(gene_count > 100)

ge.RAW.2021.day0.T.long.filter <- ge.RAW.2021.day0.T.long %>%
  dplyr::filter(gene_id %in% genecount_filter$gene_id)

# Calculate medians count per subject
medcorr <- ge.RAW.2021.day0.T.long.filter  %>%
  group_by(subject_id) %>%
  mutate(count1 = ifelse(count == 0, NA, count)) %>%
  dplyr::summarise(n=n(), 
                   median = median(count), 
                   median_exlude_0 = median(count1, na.rm = T), 
                   sum=sum(count)) %>%
  dplyr::select(median_exlude_0, subject_id)

ge.RAW.2021.day0.T.long.median <- left_join(ge.RAW.2021.day0.T.long.filter, 
                                            medcorr, by="subject_id")

# Normalize the counts
ge.RAW.2021.day0.T.long.norm <- ge.RAW.2021.day0.T.long.median %>%
   dplyr::mutate(norm_count = ((count * 10000)/median_exlude_0)/10000)

# Convert back to normal matrix format
ge.RAW.2021.day0.T.norm <- ge.RAW.2021.day0.T.long.norm %>% 
   dplyr::select(gene_id, subject_id, norm_count) %>% 
   pivot_wider(names_from = subject_id, values_from = norm_count) %>% 
   column_to_rownames("gene_id")


```

## Step 8: Make predictions for the CMI-PB 2021 data

```{r}

# Calculate the gene signature score with TPM normalized gene expression
df.gene.sig.pred.2021 <- geneSigCalc(gene.ex.data=ge.TPM.2021.day0.T, 
                                     score='missing', 
                                     gene.sig=gene.sig,
                                     genes.up=NULL, 
                                     genes.down=NULL,
                                     verbose=TRUE)

# Calculate the class probability with the Naive Bayes classifier
df.NB.pred.2021 <- mlModelPredict(gene.ex.data=ge.RAW.2021.day0.T.norm, 
                                  study='Bartholomeus_2018',
                                  post.ab.data=abt.2021.day14,
                                  pre.ab.data=abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag,
                                  model=model1)

# Read submisson format
submission.format <- fread("../../Data/2021_cmipb_challenge/2021_submission_format.tsv") %>% 
   dplyr::rename(subject_id = Subject_ID)

# HBV pre-vaccine transfer signature 2021 predictions
Bartholomeus.2018.gene.sig.2021.pred <- submission.format %>% 
   left_join(., df.gene.sig.pred.2021, by="subject_id")
Bartholomeus.2018.gene.sig.2021.pred[, c(5:11)] <- Bartholomeus.2018.gene.sig.2021.pred$rank.predictor
Bartholomeus.2018.gene.sig.2021.pred <- Bartholomeus.2018.gene.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Bartholomeus.2018.gene.sig.2021.pred, 
            file = "Results/2021_predictions/bartholomeus_2018_gene_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# HBV pre-vaccine transfer signature 2021 predictions
Bartholomeus.2018.NB.2021.pred <- submission.format %>% 
   left_join(., df.NB.pred.2021, by="subject_id")
Bartholomeus.2018.NB.2021.pred[, c(5:11)] <- Bartholomeus.2018.NB.2021.pred$rank.predictor
Bartholomeus.2018.NB.2021.pred <- Bartholomeus.2018.NB.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Bartholomeus.2018.NB.2021.pred, 
            file = "Results/2021_predictions/bartholomeus_2018_NB_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

```

## Step 9: Evaluate the 23 DEGs and the Naive Bayes classifier on the CMI-PB 2021 dataset

```{r model evaluation}

# Calculate the gene signature score with TPM normalized gene expression
df.gene.sig.2021.pred <- geneSigCalc(gene.ex.data=ge.TPM.2021.day0.T, 
                                     score='missing', 
                                     gene.sig=gene.sig,
                                     genes.up=NULL, 
                                     genes.down=NULL,
                                     post.ab.data=abt.2021.day14,
                                     pre.ab.data=abt.2021.day0,
                                     filter.ab.ag=selected.iso.ag,
                                     verbose=TRUE)

# Evaluate gene signature with TPM normalized gene expression
Bartholomeus.2018.gene.sig.2021.results <- evaluateModel(df.prediction=df.gene.sig.2021.pred, 
                                                         N.perm = 1000, direction = "missing",
                                                         add.to.title='- TPM normalized',
                                                         score.label='Differential genomic signature score',
                                                         fn1='geneSig_vs_day14_heatmap_TPM',
                                                         fn2='geneSig_vs_day14_correlation_TPM',
                                                         fn3='geneSig_vs_FC_heatmap_TPM',
                                                         fn4='geneSig_vs_FC_correlation_TPM',
                                                         fn5='geneSig_vs_day14_AUC_TPM',
                                                         fn6='geneSig_vs_day14_boxplot_TPM',
                                                         fn7='geneSig_vs_FC_AUC_TPM',
                                                         fn8='geneSig_vs_FC_boxplot_TPM',
                                                         path = 'Results/2021/Gene_signature/')

save(Bartholomeus.2018.gene.sig.2021.results, 
     file = "Results/2021/Gene_signature/Bartholomeus.2018.gene.sig.2021.results.RData")


# Calculate the class probability with the Naive Bayes classifier
df.NB.2021.pred <- mlModelPredict(gene.ex.data=ge.RAW.2021.day0.T.norm, 
                                  study='Bartholomeus_2018',
                                  post.ab.data=abt.2021.day14,
                                  pre.ab.data=abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag,
                                  model=model1)


# Evaluate the class probability predicted by the Naive Bayes classifier
Bartholomeus.2018.NB.2021.results <- evaluateModel(df.prediction=df.NB.2021.pred, 
                                                   direction='missing',
                                                   N.perm = 1000,
                                                   score.label='Class probability',
                                                   fn1='NBpred_vs_day14_heatmap',
                                                   fn2='NBpred_vs_day14_correlation',
                                                   fn3='NBpred_vs_FC_heatmap',
                                                   fn4='NBpred_vs_FC_correlation',
                                                   fn5='NBpred_vs_day14_AUC',
                                                   fn6='NBpred_vs_day14_boxplot',
                                                   fn7='NBpred_vs_FC_AUC',
                                                   fn8='NBpred_vs_FC_boxplot',
                                                   path = 'Results/2021/Naive_bayes/')
# Save results
save(Bartholomeus.2018.NB.2021.results, 
     file = "Results/2021/Naive_bayes/Bartholomeus.2018.NB.2021.results.RData")


```