---
title: "Study 1 - Multicohort analysis reveals baseline transcriptional predictors of influenza vaccination responses (PMID: 28842433)"
author: "Mikkel Niklas Rasmussen"
date: "2/15/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Applying the 9-gene signature presented in study by HIPC-CHI Signatures Project Team and HIPC-I Consortium (PMID: 28842433) on the CMI-PD dataset

```{r load libaries, echo=FALSE}

setwd(dirname(rstudioapi::getSourceEditorContext()$path))

suppressMessages(library('data.table'))
suppressMessages(library('dplyr'))
suppressMessages(library('tidyr'))
suppressMessages(library('Biobase'))
suppressMessages(library('DESeq2'))
suppressMessages(library('pROC'))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("ggplot2"))
suppressMessages(library("biomaRt"))
suppressMessages(library("qusage"))

# Load functions for evaluating gene signatures
source('../R/modelEvaluate.R')
```

#### Load the CMI-PB datasets

```{r load data}

DATA_DIR_raw <- "../Data/raw"
DATA_DIR_full <- "../Data/processed/full"

# Info table (meta data)
meta.2020 <- fread(file.path(DATA_DIR_full, "meta.2020.pivoted.tsv"))
meta.2021 <- fread(file.path(DATA_DIR_full, "meta.2021.pivoted.tsv"))
specimen.table.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_specimen.csv"))
specimen.table.2021 <- fread(file.path(DATA_DIR_raw, "2021BD_specimen.csv"))

# Antibody titer data
ab.titer.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_ab_titer.csv"))
ab.titer.2021.day0 <- fread(file.path(DATA_DIR_raw, "2021BD_ab_titer.csv"))
ab.titer.2021.day14 <- fread(file.path(DATA_DIR_raw, "../2021_ab_titer_day14.csv"))

# Gene expression data for day 0
ge.RAW.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2020.day0.pivoted.tsv"))
ge.RAW.2021.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2021.day0.pivoted.tsv"))

```

## Step 1: Load the 9-gene signature 

```{r}

# Isotype-antigens which are considered the most important
selected.iso.ag <- c("IgG-PT", "IgG-FHA",   "IgG-PRN", "IgG1-PT",
                     "IgG1-FHA", "IgG4-PT", "IgG4-FHA")

# Gene signature reported in the original manuscript to be significantly 
# increased expression in high responders
gene.sig <- c('RAB24', 'GRB2', 'DPP3', 'ACTB', 'MVP', 'DPP7', 'ARPC4', 
              'PLEKHB2', 'ARRB1')

```

## Step 2: Pre-process CMI-PB 2020 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
abt.2020.all <- ab.titer.2020 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14, 30)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data points on all days
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14) 

# Subset antibody titer
abt.2020.subset <- abt.2020.all[abt.2020.all$subject_id %in% ids, ]
abt.2020.day0 <- abt.2020.subset[abt.2020.subset$timepoint == 0, ]
abt.2020.day14 <- abt.2020.subset[abt.2020.subset$timepoint == 14, ]


```

## Step 3: Pre-process CMI-PB 2020 raw gene counts matrix with DESeq2

```{r pre-process gene data}

# Tranpose the raw cout gene expression data
ge.RAW.2020.day0.T <- ge.RAW.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2020.day0 <- meta.2020[meta.2020$planned_day_relative_to_boost == 0, ]
meta.2020.day0..mask <- meta.2020.day0$subject_id %in% colnames(ge.RAW.2020.day0.T)
meta.2020.day0 <- meta.2020.day0[meta.2020.day0..mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2020.day0.T,
                              colData = meta.2020.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2020.day0 <- assay(vsd)

```

## Step 4: Evaluate the predictive power of the gene signatures on the CMI-PB 2020 dataset

```{r model evaluation}

# 9-gene signature significantly increased expression in high responders
# Calculate the gene signature score with DESeq2 normalized gene expression
df.gene.sig.pred.2020 <- geneSigCalc(gene.ex.data=ge.norm.2020.day0, 
                                     score='geometric', 
                                     gene.sig=gene.sig, 
                                     post.ab.data=abt.2020.day14,
                                     pre.ab.data=abt.2020.day0,
                                     filter.ab.ag=selected.iso.ag,
                                     verbose=TRUE)

# Evaluate gene signature with DESeq normalized gene expression
Avey.2017.gene.sig.2020.results <- evaluateModel(df.prediction=df.gene.sig.pred.2020, 
                                                 N.perm = 10000, direction = "<",
                                                 add.to.title='- DESeq2 VST normalized',
                                                 fn1='geneSig_vs_day14_heatmap',
                                                 fn2='geneSig_vs_day14_correlation',
                                                 fn3='geneSig_vs_FC_heatmap',
                                                 fn4='geneSig_vs_FC_correlation',
                                                 fn5='geneSig_vs_day14_AUC',
                                                 fn6='geneSig_vs_day14_boxplot',
                                                 fn7='geneSig_vs_FC_AUC',
                                                 fn8='geneSig_vs_FC_boxplot',
                                                 path='Results/2020/Gene_signature/')

save(Avey.2017.gene.sig.2020.results, 
     file = "Results/2020/Gene_signature/Avey.2017.gene.sig.2020.results.RData")

```

## Step 4: Evaluate the predictive power on the CMI-PB dataset of the gene modul activity of the top 3 most significantly different between low and high responders 

```{r}
################ DESEQ normalized gene expression ################
# Calculate gene module intensity for the 'BCR signaling (M54)' module
df.module.M54.2020.pred <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0, 
                                          module='BCR signaling (M54)',
                                          score='qusage',
                                          post.ab.data=abt.2020.day14,
                                          pre.ab.data=abt.2020.day0, 
                                          filter.ab.ag=selected.iso.ag,
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'BCR signaling (M54)' module
Avey.2017.module.M54.2020.results <- evaluateModel(df.prediction=df.module.M54.2020.pred, 
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (BCR signaling (M54))',
                                                   fn1='geneModule_M54_vs_day14_heatmap_DESEQ',
                                                   fn2='geneModule_M54_vs_day14_correlation_DESEQ ',
                                                   fn3='geneModule_M54_vs_FC_heatmap_DESEQ',
                                                   fn4='geneModule_M54_vs_FC_correlation_DESEQ ',
                                                   fn5='geneModule_M54_vs_day14_AUC_DESEQ',
                                                   fn6='geneModule_M54_vs_day14_boxplot_DESEQ',
                                                   fn7='geneModule_M54_vs_FC_AUC_DESEQ',
                                                   fn8='geneModule_M54_vs_FC_boxplot_DESEQ',
                                                   path = 'Results/2020/M54/')

save(Avey.2017.module.M54.2020.results, 
     file = "Results/2020/M54/Avey.2017.module.M54.2020.results.RData")

# Calculate gene module intensity for the 'platelet activation (III) (M42)' module
df.module.M42.2020.pred <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0, 
                                          module='platelet activation (III) (M42)',
                                          score='qusage',
                                          post.ab.data=abt.2020.day14,
                                          pre.ab.data=abt.2020.day0, 
                                          filter.ab.ag=selected.iso.ag,
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'platelet activation (III) (M42)' module
Avey.2017.module.M42.2020.results <- evaluateModel(df.prediction=df.module.M42.2020.pred,
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (platelet activation (III) (M42))',
                                                   fn1='geneModule_M42_vs_day14_heatmap_DESEQ',
                                                   fn2='geneModule_M42_vs_day14_correlation_DESEQ ',
                                                   fn3='geneModule_M42_vs_FC_heatmap_DESEQ',
                                                   fn4='geneModule_M42_vs_FC_correlation_DESEQ ',
                                                   fn5='geneModule_M42_vs_day14_AUC_DESEQ',
                                                   fn6='geneModule_M42_vs_day14_boxplot_DESEQ',
                                                   fn7='geneModule_M42_vs_FC_AUC_DESEQ',
                                                   fn8='geneModule_M42_vs_FC_boxplot_DESEQ',
                                                   path = 'Results/2020/M42/')

save(Avey.2017.module.M42.2020.results, 
     file = "Results/2020/M42/Avey.2017.module.M42.2020.results.RData")


# Calculate gene module intensity for the 'inflammatory response (M33)' module
df.module.M33.2020.pred <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0, 
                                          module='inflammatory response (M33)',
                                          score='qusage',
                                          post.ab.data=abt.2020.day14,
                                          pre.ab.data=abt.2020.day0, 
                                          filter.ab.ag=selected.iso.ag, 
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'inflammatory response (M33)' module
Avey.2017.module.M33.2020.results <- evaluateModel(df.prediction=df.module.M33.2020.pred,
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (inflammatory response (M33))',
                                                   fn1='geneModule_M33_vs_day14_heatmap_DESEQ',
                                                   fn2='geneModule_M33_vs_day14_correlation_DESEQ ',
                                                   fn3='geneModule_M33_vs_FC_heatmap_DESEQ',
                                                   fn4='geneModule_M33_vs_FC_correlation_DESEQ ',
                                                   fn5='geneModule_M33_vs_day14_AUC_DESEQ',
                                                   fn6='geneModule_M33_vs_day14_boxplot_DESEQ',
                                                   fn7='geneModule_M33_vs_FC_AUC_DESEQ',
                                                   fn8='geneModule_M33_vs_FC_boxplot_DESEQ',
                                                   path = 'Results/2020/M33/')

save(Avey.2017.module.M33.2020.results, 
     file = "Results/2020/M33/Avey.2017.module.M33.2020.results.RData")
```

## Step 5: Pre-process CMI-PB 2021 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
ab.titer.day0.2021.all <- ab.titer.2021.day0 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

ab.titer.subject.ids <- unique(ab.titer.day0.2021.all$subject_id)
abt.2021.day0 <- ab.titer.day0.2021.all[ab.titer.day0.2021.all$timepoint == 0, ]
abt.2021.day0 <- abt.2021.day0 %>% 
   filter(isotype_antigen %in% selected.iso.ag)

# Change format of the day 14 antibody titers
abt.2021.day14 <- ab.titer.2021.day14 %>% 
   separate(task, into=c("task", "antigen"), sep = "\\) ") %>% 
   dplyr::mutate(isotype_antigen = recode(antigen, 
                                          `Pertussis Toxin` = "IgG-PT", 
                                          `FHA` = 'IgG-FHA', 
                                          `Pertactin` =  'IgG-PRN',
                                          `IgG1 - Pertussis toxin` = "IgG1-PT",
                                          `IgG1 - FHA` = "IgG1-FHA",
                                          `IgG4 - Pertussis toxin` = "IgG4-PT",
                                          `IgG4 - FHA` = "IgG4-FHA"))

```

## Step 6: Pre-process CMI-PB 2021 raw gene counts matrix with DESeq2

```{r pre-process gene data}

# Subset gene expression data based on subjects with antibody titers
ge.RAW.2021.day0 <- ge.RAW.2021.day0[ge.RAW.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# Transpose the raw cout gene expression data
ge.RAW.2021.day0.T <- ge.RAW.2021.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2021.day0 <- meta.2021[meta.2021$planned_day_relative_to_boost == 0, ]
meta.2021.day0..mask <- meta.2021.day0$subject_id %in% colnames(ge.RAW.2021.day0.T)
meta.2021.day0 <- meta.2021.day0[meta.2021.day0..mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2021.day0.T,
                              colData = meta.2021.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2021.day0 <- assay(vsd)

```

## Step 7: Make prediction for the CMI-PB 2021 prediction contest

```{r}

# 9-gene signature significantly increased expression in high responders
# Calculate the gene signature score with DESeq2 normalized gene expression
df.gene.sig.pred.2021 <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                     score='geometric', 
                                     gene.sig=gene.sig,
                                     verbose=TRUE)

# Calculate gene module intensity for the 'BCR signaling (M54)' module
df.module.M54.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='BCR signaling (M54)',
                                          score='qusage',
                                          verbose=TRUE)

# Calculate gene module intensity for the 'platelet activation (III) (M42)' module
df.module.M42.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='platelet activation (III) (M42)',
                                          score='qusage',
                                          verbose=TRUE)

# Calculate gene module intensity for the 'inflammatory response (M33)' module
df.module.M33.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='inflammatory response (M33)',
                                          score='qusage',
                                          verbose=TRUE)

# Read submisson format
submission.format <- fread("../Data/2021_submission_format.tsv") %>% 
   dplyr::rename(subject_id = Subject_ID)

# 9-gene signature 2021 predictions
Avey.2017.gene.sig.2021.pred <- submission.format %>% 
   left_join(., df.gene.sig.pred.2021, by="subject_id")
Avey.2017.gene.sig.2021.pred[, c(5:11)] <- Avey.2017.gene.sig.2021.pred$rank.predictor
Avey.2017.gene.sig.2021.pred <- Avey.2017.gene.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Avey.2017.gene.sig.2021.pred, 
            file = "Results/2021_predictions/avey_2017_gene_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# 'BCR signaling (M54)' BTM 2021 predictions
Avey.2017.M54.2021.pred <- submission.format %>% 
   left_join(., df.module.M54.pred.2021, by="subject_id")
Avey.2017.M54.2021.pred[, c(5:11)] <- Avey.2017.M54.2021.pred$rank.predictor
Avey.2017.M54.2021.pred <- Avey.2017.M54.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Avey.2017.M54.2021.pred, 
            file = "Results/2021_predictions/avey_2017_M54_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# 'Platelet activation (III) (M42)' BTM 2021 predictions
Avey.2017.M42.2021.pred <- submission.format %>% 
   left_join(., df.module.M42.pred.2021, by="subject_id")
Avey.2017.M42.2021.pred[, c(5:11)] <- Avey.2017.M42.2021.pred$rank.predictor
Avey.2017.M42.2021.pred <- Avey.2017.M42.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Avey.2017.M42.2021.pred, 
            file = "Results/2021_predictions/avey_2017_M42_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# 'Inflammatory response (M33)' BTM 2021 predictions
Avey.2017.M33.2021.pred <- submission.format %>% 
   left_join(., df.module.M33.pred.2021, by="subject_id")
Avey.2017.M33.2021.pred[, c(5:11)] <- Avey.2017.M33.2021.pred$rank.predictor
Avey.2017.M33.2021.pred <- Avey.2017.M33.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Avey.2017.M33.2021.pred, 
            file = "Results/2021_predictions/avey_2017_M33_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

```

## Step 8: Evaluate the performance of the gene signature predictions for the CMI-PB 2021 data

```{r}

# Calculate the gene signature score with DESeq2 normalized gene expression
df.gene.sig.pred.2021 <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                     score='geometric', 
                                     gene.sig=gene.sig,
                                     pre.ab.data = abt.2021.day0,
                                     post.ab.data = abt.2021.day14,
                                     filter.ab.ag=selected.iso.ag,
                                     verbose=TRUE)

# Evaluate gene signature with DESeq normalized gene expression
Avey.2017.gene.sig.2021.results <- evaluateModel(df.prediction=df.gene.sig.pred.2021, 
                                                 N.perm = 10000, direction = "<",
                                                 add.to.title='- DESeq2 VST normalized',
                                                 fn1='geneSig_vs_day14_heatmap',
                                                 fn2='geneSig_vs_day14_correlation',
                                                 fn3='geneSig_vs_FC_heatmap',
                                                 fn4='geneSig_vs_FC_correlation',
                                                 fn5='geneSig_vs_day14_AUC',
                                                 fn6='geneSig_vs_day14_boxplot',
                                                 fn7='geneSig_vs_FC_AUC',
                                                 fn8='geneSig_vs_FC_boxplot',
                                                 path='Results/2021/Gene_signature/')

save(Avey.2017.gene.sig.2021.results, 
     file = "Results/2021/Gene_signature/Avey.2017.gene.sig.2021.results.RData")

```

## Step 9: Evaluate the predictive power on the CMI-PB 2021 dataset of the 3 BTMs 
```{r}
################ DESEQ normalized gene expression ################
# Calculate gene module intensity for the 'BCR signaling (M54)' module
df.module.M54.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='BCR signaling (M54)',
                                          score='qusage',
                                          post.ab.data=abt.2021.day14,
                                          pre.ab.data=abt.2021.day0, 
                                          filter.ab.ag=selected.iso.ag,
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'BCR signaling (M54)' module
Avey.2017.module.M54.2021.results <- evaluateModel(df.prediction=df.module.M54.2021.pred, 
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (BCR signaling (M54))',
                                                   fn1='geneModule_M54_vs_day14_heatmap',
                                                   fn2='geneModule_M54_vs_day14_correlation',
                                                   fn3='geneModule_M54_vs_FC_heatmap',
                                                   fn4='geneModule_M54_vs_FC_correlation ',
                                                   fn5='geneModule_M54_vs_day14_AUC',
                                                   fn6='geneModule_M54_vs_day14_boxplot',
                                                   fn7='geneModule_M54_vs_FC_AUC',
                                                   fn8='geneModule_M54_vs_FC_boxplot',
                                                   path = 'Results/2021/M54/')

save(Avey.2017.module.M54.2021.results, 
     file = "Results/2021/M54/Avey.2017.module.M54.2021.results.RData")

# Calculate gene module intensity for the 'platelet activation (III) (M42)' module
df.module.M42.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='platelet activation (III) (M42)',
                                          score='qusage',
                                          post.ab.data=abt.2021.day14,
                                          pre.ab.data=abt.2021.day0, 
                                          filter.ab.ag=selected.iso.ag,
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'platelet activation (III) (M42)' module
Avey.2017.module.M42.2021.results <- evaluateModel(df.prediction=df.module.M42.2021.pred,
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (platelet activation (III) (M42))',
                                                   fn1='geneModule_M42_vs_day14_heatmap_DESEQ',
                                                   fn2='geneModule_M42_vs_day14_correlation_DESEQ ',
                                                   fn3='geneModule_M42_vs_FC_heatmap_DESEQ',
                                                   fn4='geneModule_M42_vs_FC_correlation_DESEQ ',
                                                   fn5='geneModule_M42_vs_day14_AUC_DESEQ',
                                                   fn6='geneModule_M42_vs_day14_boxplot_DESEQ',
                                                   fn7='geneModule_M42_vs_FC_AUC_DESEQ',
                                                   fn8='geneModule_M42_vs_FC_boxplot_DESEQ',
                                                   path = 'Results/2021/M42/')

save(Avey.2017.module.M42.2021.results, 
     file = "Results/2021/M42/Avey.2017.module.M42.2021.results.RData")


# Calculate gene module intensity for the 'inflammatory response (M33)' module
df.module.M33.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                          module='inflammatory response (M33)',
                                          score='qusage',
                                          post.ab.data=abt.2021.day14,
                                          pre.ab.data=abt.2021.day0, 
                                          filter.ab.ag=selected.iso.ag, 
                                          verbose=TRUE)

# Evaluate gene module intensity for the 'inflammatory response (M33)' module
Avey.2017.module.M33.2021.results <- evaluateModel(df.prediction=df.module.M33.2021.pred,
                                                   N.perm = 10000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label='Gene module intensity (inflammatory response (M33))',
                                                   fn1='geneModule_M33_vs_day14_heatmap_DESEQ',
                                                   fn2='geneModule_M33_vs_day14_correlation_DESEQ ',
                                                   fn3='geneModule_M33_vs_FC_heatmap_DESEQ',
                                                   fn4='geneModule_M33_vs_FC_correlation_DESEQ ',
                                                   fn5='geneModule_M33_vs_day14_AUC_DESEQ',
                                                   fn6='geneModule_M33_vs_day14_boxplot_DESEQ',
                                                   fn7='geneModule_M33_vs_FC_AUC_DESEQ',
                                                   fn8='geneModule_M33_vs_FC_boxplot_DESEQ',
                                                   path = 'Results/2021/M33/')

save(Avey.2017.module.M33.2021.results, 
     file = "Results/2021/M33/Avey.2017.module.M33.2021.results.RData")
```
