---
title: "Study 3 - Broad immune activation underlies shared set point signatures for vaccine responsiveness in healthy individuals and disease activity in patients with lupus (PMID: 32094927)"
author: "Mikkel Niklas Rasmussen"
date: "2/8/2022"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Applying the TGSig gene signature presented in study by Kotliarov et al. (PMID: 32094927) on the CMI-PD dataset

```{r load libaries, echo=FALSE}

PROJECT_DIR <- "/Users/mrasmussen/OneDrive - Danmarks Tekniske Universitet/Teknisk_Biomedicin/6_semester/Bachelorproject/Study_3-Kotliarov/baseline"

setwd(dirname(rstudioapi::getSourceEditorContext()$path))

suppressMessages(library('data.table'))
suppressMessages(library('dplyr'))
suppressMessages(library('tidyr'))
suppressMessages(library('Biobase'))
suppressMessages(library('DESeq2'))
suppressMessages(library('pROC'))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("ggplot2"))
suppressMessages(library("biomaRt"))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("geneExpressionFromGEO"))

# Load functions for evaluating gene signatures
source('../../R/functions/geneModelEvaluate.R')

```

#### Load the CMI-PB datasets

```{r load data}

DATA_DIR_raw <- "../Data/raw"
DATA_DIR_full <- "../Data/processed/full"

# Info table (meta data)
meta.2020 <- fread(file.path(DATA_DIR_full, "meta.2020.pivoted.tsv"))
meta.2021 <- fread(file.path(DATA_DIR_full, "meta.2021.pivoted.tsv"))
specimen.table.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_specimen.csv"))
specimen.table.2021 <- fread(file.path(DATA_DIR_raw, "2021BD_specimen.csv"))

# Antibody titer data
ab.titer.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_ab_titer.csv"))
ab.titer.2021.day0 <- fread(file.path(DATA_DIR_raw, "2021BD_ab_titer.csv"))
ab.titer.2021.day14 <- fread(file.path(DATA_DIR_raw, "../2021_ab_titer_day14.csv"))

# Gene expression data for day 0  - 2020
ge.TPM.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_tpm.2020.day0.pivoted.tsv"))
ge.RAW.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2020.day0.pivoted.tsv"))

# Gene expression data for day 0  - 2021
ge.RAW.2021.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2021.day0.pivoted.tsv"))

```

## Initializing function for obtaining the gene signatures

```{r initialize functions}

# load_sig function from original paper (PMID: 32094927)
# Source:
# https://github.com/niaid/baseline/tree/master/R/functions 

load_sig <- function(data=NULL, col=NULL, desc.order=T, ntop) {
  if(is.null(data)) stop("Please enter file name")
  if(is.null(col)) stop("Please enter a column name to rank genes")
  if(is.character(data)) {
    if(file.exists(data)) {
      df.sig = read.table(data, sep="\t", header=T, row.names=1, stringsAsFactors=F)
    } else {
      stop(sprintf("File not found - %s", data))
    }
  } else if (is.data.frame(data)){
    df.sig = data
  } else {
    stop("Unsuported data. Specify a file name or a data frame")
  }
  if (is.character(col) & length(col)==1) {
    predictor = df.sig[,names(df.sig) == col]
  } else if (is.numeric(col) & length(col)==1) {
    predictor = df.sig[,col]
  } else {
    stop("Unsupported column variable. Use single column name or number")
  }
  names(predictor) = rownames(df.sig)
  predictor = sort(predictor, decreasing = desc.order)
  return(names(predictor)[1:ntop])
}


```

## Step 1: Load gene signatures 

```{r}

# Load CD38 (TGSig) signature genes
fn.cd38.cor <- file.path(PROJECT_DIR, "generated_data", "CHI", 
                        "robust_corr_genes.txt")
TGSig.gene.sig <- load_sig(fn.cd38.cor, "cor.mean.sd.ratio", ntop=10)


# Load SLE-Sig (87 leading edge genes from the brown module)
SLE_Sig.gene.sig <- fread(file.path(PROJECT_DIR,                               "generated_data/fgsea_with_wgcna_modules/brown-leading-edge-isv050-87genes.txt"))$gene

# Load IFN-I-DCact gene signature 
fn.sig <- file.path(PROJECT_DIR, "generated_data", "signatures", "IFN26_ge_sig.txt")
IFN_I_DCact.gene.sig <- fread(fn.sig, header = F) %>% unlist(use.names=F)

```

## Step 2: Preprocess CMI-PB 2020 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
abt.2020.all <- ab.titer.2020 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14, 30)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data points on all days
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14)

# Subset antibody titer
abt.2020.subset <- abt.2020.all[abt.2020.all$subject_id %in% ids, ]
abt.2020.day0 <- abt.2020.subset[abt.2020.subset$timepoint == 0, ]
abt.2020.day14 <- abt.2020.subset[abt.2020.subset$timepoint == 14, ]

```

## Step 3: Preprocess raw CMI-PB 2020 gene expression count data with DESeq2 and transpose the TPM normalized gene expression

```{r pre-process gene data}

# Transpose the TPM normalized gene expression data
ge.TPM.2020.day0.T <- ge.TPM.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Tranpose the raw cout gene expression data
ge.RAW.2020.day0.T <- ge.RAW.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2020.day0 <- meta.2020[meta.2020$planned_day_relative_to_boost == 0, ]
meta.2020.day0.mask <- meta.2020.day0$subject_id %in% colnames(ge.RAW.2020.day0.T)
meta.2020.day0 <- meta.2020.day0[meta.2020.day0.mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2020.day0.T,
                              colData = meta.2020.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2020.day0 <- assay(vsd)

```

## Step 4: Evaluate the predictive power of the transciptional signatures 

```{r model evaluation}

# Isotype-antigens which are considered the most important
selected.iso.ag <- c("IgG-PT", "IgG-FHA",   "IgG-PRN", "IgG1-PT",
                     "IgG1-FHA", "IgG4-PT", "IgG4-FHA")

################ TPM normalized gene expression ################
#### TGSig ####
# Calculate the gene signature score 
df.TGSig.2020.pred.TPM <- geneSigCalc(gene.ex.data=ge.TPM.2020.day0.T, 
                                      score='mean',
                                      gene.sig=TGSig.gene.sig, 
                                      post.ab.data=abt.2020.day14,
                                      pre.ab.data=abt.2020.day0,
                                      filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with TPM normalized gene expression
Kotliarov.2020.TGSig.2020.results.TPM <- evaluateModel(df.prediction=df.TGSig.2020.pred.TPM,
                                                       N.perm = 1000, direction = "<",
                                                       add.to.title='- TPM normalized',
                                                       score.label = 'TGSig signature score',
                                                       fn1='TGSig_vs_day14_heatmap_TPM',
                                                       fn2='TGSig_vs_day14_correlation_TPM ',
                                                       fn3='TGSig_vs_FC_heatmap_TPM',
                                                       fn4='TGSig_vs_FC_correlation_TPM ',
                                                       fn5='TGSig_vs_day14_AUC_TPM',
                                                       fn6='TGSig_vs_day14_boxplot_TPM',
                                                       fn7='TGSig_vs_FC_AUC_TPM',
                                                       fn8='TGSig_vs_FC_boxplot_TPM',
                                                       path = 'Results/2020/TPM/TGSig/')
# Save results
save(Kotliarov.2020.TGSig.2020.results.TPM, 
     file = "Results/2020/TPM/TGSig/Kotliarov.2020.TGSig.2020.results.TPM.RData")

#### SLE-Sig ####
# Calculate the gene signature score 
df.SLE.Sig.2020.pred.TPM <- geneSigCalc(gene.ex.data=ge.TPM.2020.day0.T, 
                                        score='mean',
                                        gene.sig=SLE_Sig.gene.sig, 
                                        post.ab.data=abt.2020.day14,
                                        pre.ab.data=abt.2020.day0,
                                        filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with TPM normalized gene expression
Kotliarov.2020.SLE.Sig.2020.results.TPM <- evaluateModel(df.prediction=df.SLE.Sig.2020.pred.TPM,
                                                         N.perm = 1000, direction = "<",
                                                         add.to.title='- TPM normalized',
                                                         score.label = 'SLE-Sig signature score',
                                                         fn1='SLE-Sig_vs_day14_heatmap_TPM',
                                                         fn2='SLE-Sig_vs_day14_correlation_TPM ',
                                                         fn3='SLE-Sig_vs_FC_heatmap_TPM',
                                                         fn4='SLE-Sig_vs_FC_correlation_TPM ',
                                                         fn5='SLE-Sig_vs_day14_AUC_TPM',
                                                         fn6='SLE-Sig_vs_day14_boxplot_TPM',
                                                         fn7='SLE-Sig_vs_FC_AUC_TPM',
                                                         fn8='SLE-Sig_vs_FC_boxplot_TPM',
                                                         path = 'Results/2020/TPM/SLE-Sig/')
# Save results
save(Kotliarov.2020.SLE.Sig.2020.results.TPM, 
     file = "Results/2020/TPM/SLE-Sig/Kotliarov.2020.SLE.Sig.2020.results.TPM.RData")

#### IFN-I-DCact ####
# Calculate the gene signature score 
df.IFN.I.DCact.2020.pred.TPM <- geneSigCalc(gene.ex.data=ge.TPM.2020.day0.T, 
                                            score='mean',
                                            gene.sig=IFN_I_DCact.gene.sig, 
                                            post.ab.data=abt.2020.day14,
                                            pre.ab.data=abt.2020.day0,
                                            filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with TPM normalized gene expression
Kotliarov.2020.IFN.I.DCact.2020.results.TPM <- evaluateModel(df.prediction=df.IFN.I.DCact.2020.pred.TPM,
                                                             N.perm = 1000, direction = "<",
                                                             add.to.title='- TPM normalized',
                                                             score.label = 'IFN-I-DCact signature score',
                                                             fn1='IFN-I-DCact_vs_day14_heatmap_TPM',
                                                             fn2='IFN-I-DCact_vs_day14_correlation_TPM ',
                                                             fn3='IFN-I-DCact_vs_FC_heatmap_TPM',
                                                             fn4='IFN-I-DCact_vs_FC_correlation_TPM ',
                                                             fn5='IFN-I-DCact_vs_day14_AUC_TPM',
                                                             fn6='IFN-I-DCact_vs_day14_boxplot_TPM',
                                                             fn7='IFN-I-DCact_vs_FC_AUC_TPM',
                                                             fn8='IFN-I-DCact_vs_FC_boxplot_TPM',
                                                             path = 'Results/2020/TPM/IFN-I-DCact/')
# Save results
save(Kotliarov.2020.IFN.I.DCact.2020.results.TPM, 
     file = "Results/2020/TPM/IFN-I-DCact/Kotliarov.2020.IFN.I.DCact.2020.results.TPM.RData")

################ DESeq2 normalized gene expression ###############
#### TGSig ####
# Calculate the gene signature score with DESeq2 normalized gene expression
df.TGSig.2020.pred.DESEQ <- geneSigCalc(gene.ex.data=ge.norm.2020.day0, 
                                        score='mean',
                                        gene.sig=TGSig.gene.sig, 
                                        post.ab.data=abt.2020.day14,
                                        pre.ab.data=abt.2020.day0,
                                        filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with DESeq2 normalized gene expression
Kotliarov.2020.TGSig.2020.results.DESEQ <- evaluateModel(df.prediction=df.TGSig.2020.pred.DESEQ,
                                                         N.perm = 1000, direction = "<",
                                                         add.to.title='- DESeq2 VST normalized',
                                                         score.label = 'TGSig signature score',
                                                         fn1='TGSig_vs_day14_heatmap_DESEQ',
                                                         fn2='TGSig_vs_day14_correlation_DESEQ ',
                                                         fn3='TGSig_vs_FC_heatmap_DESEQ',
                                                         fn4='TGSig_vs_FC_correlation_DESEQ ',
                                                         fn5='TGSig_vs_day14_AUC_DESEQ',
                                                         fn6='TGSig_vs_day14_boxplot_DESEQ',
                                                         fn7='TGSig_vs_FC_AUC_DESEQ',
                                                         fn8='TGSig_vs_FC_boxplot_DESEQ',
                                                         path='Results/2020/DESEQ/TGSig/')
# Save results
save(Kotliarov.2020.TGSig.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/TGSig/Kotliarov.2020.TGSig.2020.results.DESEQ.RData")

#### SLE-Sig ####
# Calculate the gene signature score 
df.SLE.Sig.2020.pred.DESEQ <- geneSigCalc(gene.ex.data=ge.norm.2020.day0, 
                                          score='mean',
                                          gene.sig=SLE_Sig.gene.sig, 
                                          post.ab.data=abt.2020.day14,
                                          pre.ab.data=abt.2020.day0,
                                          filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with DESEQ normalized gene expression
Kotliarov.2020.SLE.Sig.2020.results.DESEQ <- evaluateModel(df.prediction=df.SLE.Sig.2020.pred.DESEQ,
                                                           N.perm = 1000, direction = "<",
                                                           add.to.title='- DESeq2 VST normalized',
                                                           score.label = 'SLE-Sig signature score',
                                                           fn1='SLE-Sig_vs_day14_heatmap_DESEQ',
                                                           fn2='SLE-Sig_vs_day14_correlation_DESEQ ',
                                                           fn3='SLE-Sig_vs_FC_heatmap_DESEQ',
                                                           fn4='SLE-Sig_vs_FC_correlation_DESEQ ',
                                                           fn5='SLE-Sig_vs_day14_AUC_DESEQ',
                                                           fn6='SLE-Sig_vs_day14_boxplot_DESEQ',
                                                           fn7='SLE-Sig_vs_FC_AUC_DESEQ',
                                                           fn8='SLE-Sig_vs_FC_boxplot_DESEQ',
                                                           path = 'Results/2020/DESEQ/SLE-Sig/')
# Save results
save(Kotliarov.2020.SLE.Sig.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/SLE-Sig/Kotliarov.2020.SLE.Sig.2020.results.DESEQ.RData")

#### IFN-I-DCact ####
# Calculate the gene signature score 
df.IFN.I.DCact.2020.pred.DESEQ <- geneSigCalc(gene.ex.data=ge.norm.2020.day0, 
                                              score='mean',
                                              gene.sig=IFN_I_DCact.gene.sig, 
                                              post.ab.data=abt.2020.day14,
                                              pre.ab.data=abt.2020.day0,
                                              filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with DESEQ normalized gene expression
Kotliarov.2020.IFN.I.DCact.2020.results.DESEQ <- evaluateModel(df.prediction=df.IFN.I.DCact.2020.pred.DESEQ,
                                                               N.perm = 1000, direction = "<",
                                                               add.to.title='- DESeq2 VST normalized',
                                                               score.label = 'IFN-I-DCact signature score',
                                                               fn1='IFN-I-DCact_vs_day14_heatmap_DESEQ',
                                                               fn2='IFN-I-DCact_vs_day14_correlation_DESEQ ',
                                                               fn3='IFN-I-DCact_vs_FC_heatmap_DESEQ',
                                                               fn4='IFN-I-DCact_vs_FC_correlation_DESEQ ',
                                                               fn5='IFN-I-DCact_vs_day14_AUC_DESEQ',
                                                               fn6='IFN-I-DCact_vs_day14_boxplot_DESEQ',
                                                               fn7='IFN-I-DCact_vs_FC_AUC_DESEQ',
                                                               fn8='IFN-I-DCact_vs_FC_boxplot_DESEQ',
                                                               path = 'Results/2020/DESEQ/IFN-I-DCact/')
# Save results
save(Kotliarov.2020.IFN.I.DCact.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/IFN-I-DCact/Kotliarov.2020.IFN.I.DCact.2020.results.DESEQ.RData")

```

## Step 5: Pre-process CMI-PB 2021 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
ab.titer.day0.2021.all <- ab.titer.2021.day0 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

ab.titer.subject.ids <- unique(ab.titer.day0.2021.all$subject_id)
abt.2021.day0 <- ab.titer.day0.2021.all[ab.titer.day0.2021.all$timepoint == 0, ]
abt.2021.day0 <- abt.2021.day0 %>% 
   filter(isotype_antigen %in% selected.iso.ag)

# Change format of the day 14 antibody titers
abt.2021.day14 <- ab.titer.2021.day14 %>% 
   separate(task, into=c("task", "antigen"), sep = "\\) ") %>% 
   dplyr::mutate(isotype_antigen = recode(antigen, 
                                          `Pertussis Toxin` = "IgG-PT", 
                                          `FHA` = 'IgG-FHA', 
                                          `Pertactin` =  'IgG-PRN',
                                          `IgG1 - Pertussis toxin` = "IgG1-PT",
                                          `IgG1 - FHA` = "IgG1-FHA",
                                          `IgG4 - Pertussis toxin` = "IgG4-PT",
                                          `IgG4 - FHA` = "IgG4-FHA"))

```

## Step 6: Pre-process CMI-PB 2021 raw gene counts matrix with DESeq2

```{r pre-process gene data}

# Subset gene expression data based on subjects with antibody titers
ge.RAW.2021.day0 <- ge.RAW.2021.day0[ge.RAW.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# Tranpose the raw cout gene expression data
ge.RAW.2021.day0.T <- ge.RAW.2021.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2021.day0 <- meta.2021[meta.2021$planned_day_relative_to_boost == 0, ]
meta.2021.day0..mask <- meta.2021.day0$subject_id %in% colnames(ge.RAW.2021.day0.T)
meta.2021.day0 <- meta.2021.day0[meta.2021.day0..mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2021.day0.T,
                              colData = meta.2021.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2021.day0 <- assay(vsd)

```

## Step 7: Make prediction for CMI-PB 2021 dataset

```{r}

## TGSig
# Calculate the gene signature score with DESeq2 normalized gene expression
df.TGSig.pred.2021 <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                  score='mean',
                                  gene.sig=TGSig.gene.sig,
                                  verbose = TRUE)

## SLE-Sig
# Calculate the gene signature score 
df.SLE.Sig.pred.2021 <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                    score='mean',
                                    gene.sig=SLE_Sig.gene.sig,
                                    verbose = TRUE)

#### IFN-I-DCact ####
# Calculate the gene signature score 
df.IFN.I.DCact.pred.2021 <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                        score='mean',
                                        gene.sig=IFN_I_DCact.gene.sig,
                                        verbose = TRUE)


# Read submisson format
submission.format <- fread("../../Data/2021_cmipb_challenge/2021_submission_format.tsv") %>% 
   dplyr::rename(subject_id = Subject_ID)

# TGSig predictions
Kotliarov.2020.TGSig.2021.pred <- submission.format %>% 
   left_join(., df.TGSig.pred.2021, by="subject_id")
Kotliarov.2020.TGSig.2021.pred[, c(5:11)] <- Kotliarov.2020.TGSig.2021.pred$rank.predictor
Kotliarov.2020.TGSig.2021.pred <- Kotliarov.2020.TGSig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Kotliarov.2020.TGSig.2021.pred, 
            file = "Results/2021_predictions/kotliarov_2020_TGSig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# SLE-Sig predictions
Kotliarov.2020.SLE.Sig.2021.pred <- submission.format %>% 
   left_join(., df.SLE.Sig.pred.2021, by="subject_id")
Kotliarov.2020.SLE.Sig.2021.pred[, c(5:11)] <- Kotliarov.2020.SLE.Sig.2021.pred$rank.predictor
Kotliarov.2020.SLE.Sig.2021.pred <- Kotliarov.2020.SLE.Sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Kotliarov.2020.SLE.Sig.2021.pred, 
            file = "Results/2021_predictions/kotliarov_2020_SLE-Sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# 'Platelet activation (III) (M42)' BTM 2021 predictions
Kotliarov.2020.IFN.I.DCact.2021.pred <- submission.format %>% 
   left_join(., df.IFN.I.DCact.pred.2021, by="subject_id")
Kotliarov.2020.IFN.I.DCact.2021.pred[, c(5:11)] <- Kotliarov.2020.IFN.I.DCact.2021.pred$rank.predictor
Kotliarov.2020.IFN.I.DCact.2021.pred <- Kotliarov.2020.IFN.I.DCact.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Kotliarov.2020.IFN.I.DCact.2021.pred, 
            file = "Results/2021_predictions/kotliarov_2020_IFN-I-DCact_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

```

## Step 8: Evaluate the performance of the gene signatures predictions for the CMI-PB 2021 data

```{r}

## TGSig
# Calculate the gene signature score with DESeq2 normalized gene expression
df.TGSig.2021.pred <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                  score='mean',
                                  gene.sig=TGSig.gene.sig,
                                  post.ab.data = abt.2021.day14,
                                  pre.ab.data = abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag,
                                  verbose = TRUE)

# Evaluate gene signature with DESeq2 normalized gene expression
Kotliarov.2020.TGSig.2021.results <- evaluateModel(df.prediction=df.TGSig.2021.pred,
                                                   N.perm = 1000, direction = "<",
                                                   add.to.title='- DESeq2 VST normalized',
                                                   score.label = 'TGSig signature score',
                                                   fn1='TGSig_vs_day14_heatmap',
                                                   fn2='TGSig_vs_day14_correlation',
                                                   fn3='TGSig_vs_FC_heatmap',
                                                   fn4='TGSig_vs_FC_correlation',
                                                   fn5='TGSig_vs_day14_AUC',
                                                   fn6='TGSig_vs_day14_boxplot',
                                                   fn7='TGSig_vs_FC_AUC',
                                                   fn8='TGSig_vs_FC_boxplot',
                                                   path='Results/2021/TGSig/')
# Save results
save(Kotliarov.2020.TGSig.2021.results, 
     file = "Results/2021/TGSig/Kotliarov.2020.TGSig.2021.results.RData")

#### SLE-Sig ####
# Calculate the gene signature score 
df.SLE.Sig.2021.pred <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                    score='mean',
                                    gene.sig=SLE_Sig.gene.sig, 
                                    post.ab.data=abt.2021.day14,
                                    pre.ab.data=abt.2021.day0,
                                    filter.ab.ag=selected.iso.ag,
                                    verbose=TRUE)

# Evaluate gene signature with DESEQ normalized gene expression
Kotliarov.2020.SLE.Sig.2021.results <- evaluateModel(df.prediction=df.SLE.Sig.2021.pred,
                                                     N.perm = 1000, direction = "<",
                                                     add.to.title='- DESeq2 VST normalized',
                                                     score.label = 'SLE-Sig signature score',
                                                     fn1='SLE-Sig_vs_day14_heatmap',
                                                     fn2='SLE-Sig_vs_day14_correlation',
                                                     fn3='SLE-Sig_vs_FC_heatmap',
                                                     fn4='SLE-Sig_vs_FC_correlation',
                                                     fn5='SLE-Sig_vs_day14_AUC',
                                                     fn6='SLE-Sig_vs_day14_boxplot',
                                                     fn7='SLE-Sig_vs_FC_AUC',
                                                     fn8='SLE-Sig_vs_FC_boxplot',
                                                     path = 'Results/2021/SLE-Sig/')
# Save results
save(Kotliarov.2020.SLE.Sig.2021.results, 
     file = "Results/2021/SLE-Sig/Kotliarov.2020.SLE.Sig.2021.results.RData")

#### IFN-I-DCact ####
# Calculate the gene signature score 
df.IFN.I.DCact.2021.pred <- geneSigCalc(gene.ex.data=ge.norm.2021.day0, 
                                        score='mean',
                                        gene.sig=IFN_I_DCact.gene.sig, 
                                        post.ab.data=abt.2021.day14,
                                        pre.ab.data=abt.2021.day0,
                                        filter.ab.ag=selected.iso.ag,
                                        verbose=TRUE)

# Evaluate gene signature with DESEQ normalized gene expression
Kotliarov.2020.IFN.I.DCact.2021.results <- evaluateModel(df.prediction=df.IFN.I.DCact.2021.pred,
                                                         N.perm = 1000, direction = "<",
                                                         add.to.title='- DESeq2 VST normalized',
                                                         score.label = 'IFN-I-DCact signature score',
                                                         fn1='IFN-I-DCact_vs_day14_heatmap',
                                                         fn2='IFN-I-DCact_vs_day14_correlation',
                                                         fn3='IFN-I-DCact_vs_FC_heatmap',
                                                         fn4='IFN-I-DCact_vs_FC_correlation',
                                                         fn5='IFN-I-DCact_vs_day14_AUC',
                                                         fn6='IFN-I-DCact_vs_day14_boxplot',
                                                         fn7='IFN-I-DCact_vs_FC_AUC',
                                                         fn8='IFN-I-DCact_vs_FC_boxplot',
                                                         path = 'Results/2021/IFN-I-DCact/')
# Save results
save(Kotliarov.2020.IFN.I.DCact.2021.results, 
     file = "Results/2021/IFN-I-DCact/Kotliarov.2020.IFN.I.DCact.2021.results.RData")

```

