---
title: "Study 6 - Transfer transcriptomic signatures for infectious diseases (PMID: 34031243)"
author: "Mikkel Niklas Rasmussen"
date: "4/19/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Applying the transfer signatures presented in the paper by Iulio et al. (PMID: 34031243) on the CMI-PD dataset

```{r load libaries, echo=FALSE}
setwd(dirname(rstudioapi::getSourceEditorContext()$path))

suppressMessages(library('data.table'))
suppressMessages(library('dplyr'))
suppressMessages(library('plyr'))
suppressMessages(library('tidyr'))
suppressMessages(library('Biobase'))
suppressMessages(library('DESeq2'))
suppressMessages(library('pROC'))
suppressMessages(library("ggplot2"))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("biomaRt"))
suppressMessages(library('klaR'))
suppressMessages(library("tibble"))
suppressMessages(library("readxl"))

# Load functions for evaluating gene signatures
source('../R/modelEvaluate.R')

```

#### Load the CMI-PB datasets

```{r load data}

DATA_DIR_raw <- "../Data/raw"
DATA_DIR_full <- "../Data/processed/full"

# Info table (meta data)
meta.2020 <- fread(file.path(DATA_DIR_full, "meta.2020.pivoted.tsv"))
meta.2021 <- fread(file.path(DATA_DIR_full, "meta.2021.pivoted.tsv"))
specimen.table.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_specimen.csv"))
specimen.table.2021 <- fread(file.path(DATA_DIR_raw, "2021BD_specimen.csv"))

# Antibody titer data
ab.titer.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_ab_titer.csv"))
ab.titer.2021.day0 <- fread(file.path(DATA_DIR_raw, "2021BD_ab_titer.csv"))
ab.titer.2021.day14 <- fread(file.path(DATA_DIR_raw, "../2021_ab_titer_day14.csv"))

# Gene expression data for day 0  - 2020
ge.TPM.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_tpm.2020.day0.pivoted.tsv"))
ge.RAW.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2020.day0.pivoted.tsv"))

# Gene expression data for day 0  - 2021
ge.RAW.2021.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2021.day0.pivoted.tsv"))

```

## Step 1: Load transfer signatures

```{r}

# Load transfer signatures
transfer.sig.table <- read_excel("../pnas.2022486118.sd01.xlsx",
                                 sheet = "transfer signature gene lists")

transfer.sig.list <- transfer.sig.table %>% 
  split(.$`Training subdataset Name`)

```

## Step 2: Preprocess CMI-PB antibody titer data

```{r}

#### Antibody titer data
# Normalized antibody titer data by baseline median
abt.2020.all <- ab.titer.2020 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14, 30)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data day 0 and 14
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14)

# Subset antibody titer
abt.2020.subset <- abt.2020.all[abt.2020.all$subject_id %in% ids, ]
abt.2020.day0 <- abt.2020.subset[abt.2020.subset$timepoint == 0, ]
abt.2020.day14 <- abt.2020.subset[abt.2020.subset$timepoint == 14, ]

```

## Step 3: Preprocessing of the CMI-PB 2020 gene expression data as described in orignal manuscript

```{r}

# Transpose TPM normalized gene expression
ge.TPM.2020.day0.T <- ge.TPM.2020.day0 %>% 
   column_to_rownames("subject_id") %>% 
   t()

# Pre-processing of the RNA seq gene expression is performed as described in the
# original code: 
# https://github.com/virbio/manuscript-transfer-signatures/blob/main/step1_signature_evaluation_in_training_set.r

# First the gene expression data is transformed to fit the expected format 
count_matrix <- ge.RAW.2020.day0 %>% 
   column_to_rownames("subject_id") %>% 
   t()

# IF GENE EXPRESSION MATRIX COMES FROM RNASEQ EXPERIMENT:
# load the raw counts gene expression matrix as count_matrix 
# (where rows are genes/ENSG and columns are samples)

# keep only the genes that had 20 reads in at least 10 percent of the samples
count_matrix <- count_matrix[apply(count_matrix, 1, 
                                  function(x){quantile(x, 0.9)}) > 20,]

# remove the the dot and anything coming after the dot in the ENSG gene 
# nomenclature
rownames(count_matrix) <- gsub("\\..*", "", rownames(count_matrix))

# compute the read per million
rpm <- t(t(count_matrix) / (colSums(count_matrix)/1e6))

# alternatively the million reads could have been computed before removing 
# the noisy genes...
MATCOUNTS <- log10(rpm + 1e-7) # add a pseudocount and use the log10

```

## Step 3: Evaluate the predictive power of the transfer signatures on day 14 antibody titers and fold-changes

```{r}

# Isotype-antigens which are considered the most important
selected.iso.ag <- c("IgG-PT", "IgG-FHA",   "IgG-PRN", "IgG1-PT",
                     "IgG1-FHA", "IgG4-PT", "IgG4-FHA")

# Calculate the gene signature scores for the 4 selected transfer signatures
selected.studies <- c("HBV pre-vaccine", "Influenza pre-vaccine F", 
                      "Influenza pre-vaccine M", "TB pre-vaccine")

results <- list()
i <- 1
for(study in selected.studies){
   predictions <- geneSigCalc(gene.ex.data=MATCOUNTS, 
                              ensembl.sig=transfer.sig.list[[study]]$ENSG, 
                              score='missing',
                              post.ab.data=abt.2020.day14,
                              pre.ab.data=abt.2020.day0,
                              filter.ab.ag=selected.iso.ag,
                              verbose=TRUE)
   
   # Evaluate transfer signatures
   tmp.results <- evaluateModel(df.prediction=predictions, direction='missing',
                            N.perm = 1000,
                            score.label='Gene signature score',
                            add.to.title=paste(" - ", study),
                            fn1='transferSig_vs_day14_heatmap',
                            fn2='transferSig_vs_day14_correlation',
                            fn3='transferSig_vs_FC_heatmap',
                            fn4='transferSig_vs_FC_correlation',
                            fn5='transferSig_vs_day14_AUC',
                            fn6='transferSig_vs_day14_boxplot',
                            fn7='transferSig_vs_FC_AUC',
                            fn8='transferSig_vs_FC_boxplot',
                            path = paste0('Results/2020/', study, "/"))
   
   results <- append(results, list(tmp.results))
   names(results)[[i]] <- study 
   i <- i + 1

}

# Save results
Iulio.2021.HBV.transfer.sig.2020.results <- results[["HBV pre-vaccine"]]
save(Iulio.2021.HBV.transfer.sig.2020.results, 
     file="Results/2020/HBV pre-vaccine/Iulio.2021.HBV.transfer.sig.2020.results.RData")

Iulio.2021.Inf.F.transfer.sig.2020.results <- results[["Influenza pre-vaccine F"]]
save(Iulio.2021.Inf.F.transfer.sig.2020.results, 
     file="Results/2020/Influenza pre-vaccine F/Iulio.2021.Inf.F.transfer.sig.2020.results.RData")

Iulio.2021.Inf.M.transfer.sig.2020.results <- results[["Influenza pre-vaccine M"]]
save(Iulio.2021.Inf.M.transfer.sig.2020.results, 
     file="Results/2020/Influenza pre-vaccine M/Iulio.2021.Inf.M.transfer.sig.2020.results.RData")

Iulio.2021.TB.transfer.sig.2020.results <- results[["TB pre-vaccine"]]
save(Iulio.2021.TB.transfer.sig.2020.results, 
     file="Results/2020/TB pre-vaccine/Iulio.2021.TB.transfer.sig.2020.results.RData")

```

## Step 4: Pre-process CMI-PB 2021 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
ab.titer.day0.2021.all <- ab.titer.2021.day0 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

ab.titer.subject.ids <- unique(ab.titer.day0.2021.all$subject_id)
abt.2021.day0 <- ab.titer.day0.2021.all[ab.titer.day0.2021.all$timepoint == 0, ]
abt.2021.day0 <- abt.2021.day0 %>% 
   filter(isotype_antigen %in% selected.iso.ag)

# Change format of the day 14 antibody titers
abt.2021.day14 <- ab.titer.2021.day14 %>% 
   separate(task, into=c("task", "antigen"), sep = "\\) ") %>% 
   dplyr::mutate(isotype_antigen = recode(antigen, 
                                          `Pertussis Toxin` = "IgG-PT", 
                                          `FHA` = 'IgG-FHA', 
                                          `Pertactin` =  'IgG-PRN',
                                          `IgG1 - Pertussis toxin` = "IgG1-PT",
                                          `IgG1 - FHA` = "IgG1-FHA",
                                          `IgG4 - Pertussis toxin` = "IgG4-PT",
                                          `IgG4 - FHA` = "IgG4-FHA"))


```

## Step 5: Preprocessing of the CMI-PB 2021 gene expression data as described in orignal manuscript

```{r}
# Subset gene expression data based on subjects with antibody titers
ge.RAW.2021.day0 <- ge.RAW.2021.day0[ge.RAW.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# First the gene expression data is transformed to fit the expected format 
count_matrix <- ge.RAW.2021.day0 %>% 
   column_to_rownames("subject_id") %>% 
   t()

# Pre-processing of the RNA seq gene expression is performed as described in the
# original code: 
# https://github.com/virbio/manuscript-transfer-signatures/blob/main/step1_signature_evaluation_in_training_set.r

# IF GENE EXPRESSION MATRIX COMES FROM RNASEQ EXPERIMENT:
# load the raw counts gene expression matrix as count_matrix 
# (where rows are genes/ENSG and columns are samples)

# keep only the genes that had 20 reads in at least 10 percent of the samples
count_matrix <- count_matrix[apply(count_matrix, 1, 
                                  function(x){quantile(x, 0.9)}) > 20,]

# remove the the dot and anything coming after the dot in the ENSG gene 
# nomenclature
rownames(count_matrix) <- gsub("\\..*", "", rownames(count_matrix))

# compute the read per million
rpm <- t(t(count_matrix) / (colSums(count_matrix)/1e6))

# alternatively the million reads could have been computed before removing 
# the noisy genes...
MATCOUNTS.2021 <- log10(rpm + 1e-7) # add a pseudocount and use the log10

```

## Step 6: Make prediction for CMI-PB 2021 dataset

```{r}

# HBV pre-vaccine transfer signature
df.HBV.pre.vac.pred.2021 <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                        ensembl.sig=transfer.sig.list[["HBV pre-vaccine"]]$ENSG, 
                                        score='missing',
                                        verbose=TRUE)

# Influenza pre-vaccine Female dataset 
Influenza.pre.vac.F.pred.2021 <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                             ensembl.sig=transfer.sig.list[["Influenza pre-vaccine F"]]$ENSG, 
                                             score='missing',
                                             verbose=TRUE)

# Influenza pre-vaccine Male dataset 
Influenza.pre.vac.M.pred.2021 <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                             ensembl.sig=transfer.sig.list[["Influenza pre-vaccine M"]]$ENSG,   
                                             score='missing',
                                             verbose=TRUE)

# TB pre-vaccine dataset
TB.pre.vac.pred.2021 <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                    ensembl.sig=transfer.sig.list[["TB pre-vaccine"]]$ENSG, 
                                    score='missing',
                                    verbose=TRUE)

# Read submisson format
submission.format <- fread("../../Data/2021_cmipb_challenge/2021_submission_format.tsv") %>% 
   dplyr::rename(subject_id = Subject_ID)

# HBV pre-vaccine transfer signature 2021 predictions
Iulio.2021.HBV.transfer.sig.2021.pred <- submission.format %>% 
   left_join(., df.HBV.pre.vac.pred.2021, by="subject_id")
Iulio.2021.HBV.transfer.sig.2021.pred[, c(5:11)] <- Iulio.2021.HBV.transfer.sig.2021.pred$rank.predictor
Iulio.2021.HBV.transfer.sig.2021.pred <- Iulio.2021.HBV.transfer.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Iulio.2021.HBV.transfer.sig.2021.pred, 
            file = "Results/2021_predictions/iulio_2021_HBV_transfer_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# Influenza pre-vaccine Female 2021 predictions
Iulio.2021.Inf.F.transfer.sig.2021.pred <- submission.format %>% 
   left_join(., Influenza.pre.vac.F.pred.2021, by="subject_id")
Iulio.2021.Inf.F.transfer.sig.2021.pred[, c(5:11)] <- Iulio.2021.Inf.F.transfer.sig.2021.pred$rank.predictor
Iulio.2021.Inf.F.transfer.sig.2021.pred <- Iulio.2021.Inf.F.transfer.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Iulio.2021.Inf.F.transfer.sig.2021.pred, 
            file = "Results/2021_predictions/iulio_2021_Inf_F_transfer_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# Influenza pre-vaccine Female 2021 predictions
Iulio.2021.Inf.M.transfer.sig.2021.pred <- submission.format %>% 
   left_join(., Influenza.pre.vac.M.pred.2021, by="subject_id")
Iulio.2021.Inf.M.transfer.sig.2021.pred[, c(5:11)] <- Iulio.2021.Inf.M.transfer.sig.2021.pred$rank.predictor
Iulio.2021.Inf.M.transfer.sig.2021.pred <- Iulio.2021.Inf.M.transfer.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Iulio.2021.Inf.M.transfer.sig.2021.pred, 
            file = "Results/2021_predictions/iulio_2021_Inf_M_transfer_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# Influenza pre-vaccine Female 2021 predictions
Iulio.2021.TB.transfer.sig.2021.pred <- submission.format %>% 
   left_join(., TB.pre.vac.pred.2021, by="subject_id")
Iulio.2021.TB.transfer.sig.2021.pred[, c(5:11)] <- Iulio.2021.TB.transfer.sig.2021.pred$rank.predictor
Iulio.2021.TB.transfer.sig.2021.pred <- Iulio.2021.TB.transfer.sig.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Iulio.2021.TB.transfer.sig.2021.pred, 
            file = "Results/2021_predictions/iulio_2021_TB_transfer_sig_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)


```

## Step 7: Evaluate the transfer signatures on the CMI-PB 2021 dataset

```{r model evaluation}

# HBV pre-vaccine transfer signature
df.HBV.pre.vac.2021.pred <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                        ensembl.sig=transfer.sig.list[["HBV pre-vaccine"]]$ENSG, 
                                        score='missing',
                                        post.ab.data=abt.2021.day14,
                                        pre.ab.data=abt.2021.day0,
                                        filter.ab.ag=selected.iso.ag,
                                        verbose=TRUE)

# Evaluate HBV pre-vaccine transfer signature
Iulio.2021.HBV.transfer.sig.2021.results <- evaluateModel(df.prediction=df.HBV.pre.vac.2021.pred, 
                                                          direction='missing',
                                                          N.perm = 1000,
                                                          score.label='Gene signature score',
                                                          add.to.title=paste(" - HBV pre-vaccine"),
                                                          fn1='transferSig_vs_day14_heatmap',
                                                          fn2='transferSig_vs_day14_correlation',
                                                          fn3='transferSig_vs_FC_heatmap',
                                                          fn4='transferSig_vs_FC_correlation',
                                                          fn5='transferSig_vs_day14_AUC',
                                                          fn6='transferSig_vs_day14_boxplot',
                                                          fn7='transferSig_vs_FC_AUC',
                                                          fn8='transferSig_vs_FC_boxplot',
                                                          path = paste0("Results/2021/HBV pre-vaccine/"))

# Save results
save(Iulio.2021.HBV.transfer.sig.2021.results, 
     file="Results/2021/HBV pre-vaccine/Iulio.2021.HBV.transfer.sig.2021.results.RData")

# Influenza pre-vaccine Female dataset 
df.Influenza.pre.vac.F.2021.pred <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                                ensembl.sig=transfer.sig.list[["Influenza pre-vaccine F"]]$ENSG, 
                                                score='missing',
                                                post.ab.data=abt.2021.day14,
                                                pre.ab.data=abt.2021.day0,
                                                filter.ab.ag=selected.iso.ag,
                                                verbose=TRUE)

# Evaluate the Influenza pre-vaccine Female transfer signature
Iulio.2021.Inf.F.transfer.sig.2021.results <- evaluateModel(df.prediction=df.Influenza.pre.vac.F.2021.pred, 
                                                            direction='missing',
                                                            N.perm = 1000,
                                                            score.label='Gene signature score',
                                                            add.to.title=paste(" - Influenza pre-vaccine F"),
                                                            fn1='transferSig_vs_day14_heatmap',
                                                            fn2='transferSig_vs_day14_correlation',
                                                            fn3='transferSig_vs_FC_heatmap',
                                                            fn4='transferSig_vs_FC_correlation',
                                                            fn5='transferSig_vs_day14_AUC',
                                                            fn6='transferSig_vs_day14_boxplot',
                                                            fn7='transferSig_vs_FC_AUC',
                                                            fn8='transferSig_vs_FC_boxplot',
                                                            path = paste0("Results/2021/Influenza pre-vaccine F/"))

# Save results
save(Iulio.2021.Inf.F.transfer.sig.2021.results, 
     file="Results/2021/Influenza pre-vaccine F/Iulio.2021.Inf.F.transfer.sig.2021.results.RData")

# Influenza pre-vaccine Male dataset 
Influenza.pre.vac.M.2021.pred <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                             ensembl.sig=transfer.sig.list[["Influenza pre-vaccine M"]]$ENSG,   
                                             score='missing',
                                             post.ab.data=abt.2021.day14,
                                             pre.ab.data=abt.2021.day0,
                                             filter.ab.ag=selected.iso.ag,
                                             verbose=TRUE)

# Evaluate the Influenza pre-vaccine Male transfer signature
Iulio.2021.Inf.M.transfer.sig.2021.results <- evaluateModel(df.prediction=Influenza.pre.vac.M.2021.pred, 
                                                            direction='missing',
                                                            N.perm = 1000,
                                                            score.label='Gene signature score',
                                                            add.to.title=paste(" - Influenza pre-vaccine M"),
                                                            fn1='transferSig_vs_day14_heatmap',
                                                            fn2='transferSig_vs_day14_correlation',
                                                            fn3='transferSig_vs_FC_heatmap',
                                                            fn4='transferSig_vs_FC_correlation',
                                                            fn5='transferSig_vs_day14_AUC',
                                                            fn6='transferSig_vs_day14_boxplot',
                                                            fn7='transferSig_vs_FC_AUC',
                                                            fn8='transferSig_vs_FC_boxplot',
                                                            path = paste0("Results/2021/Influenza pre-vaccine M/"))

# Save results
save(Iulio.2021.Inf.M.transfer.sig.2021.results, 
     file="Results/2021/Influenza pre-vaccine M/Iulio.2021.Inf.M.transfer.sig.2021.results.RData")

# TB pre-vaccine dataset
TB.pre.vac.2021.pred <- geneSigCalc(gene.ex.data=MATCOUNTS.2021, 
                                    ensembl.sig=transfer.sig.list[["TB pre-vaccine"]]$ENSG, 
                                    score='missing',
                                    post.ab.data=abt.2021.day14,
                                    pre.ab.data=abt.2021.day0,
                                    filter.ab.ag=selected.iso.ag,
                                    verbose=TRUE)

# Evaluate the TB pre-vaccine transfer signature
Iulio.2021.TB.transfer.sig.2021.results <- evaluateModel(df.prediction=TB.pre.vac.2021.pred, 
                                                            direction='missing',
                                                            N.perm = 1000,
                                                            score.label='Gene signature score',
                                                            add.to.title=paste(" - TB pre-vaccine"),
                                                            fn1='transferSig_vs_day14_heatmap',
                                                            fn2='transferSig_vs_day14_correlation',
                                                            fn3='transferSig_vs_FC_heatmap',
                                                            fn4='transferSig_vs_FC_correlation',
                                                            fn5='transferSig_vs_day14_AUC',
                                                            fn6='transferSig_vs_day14_boxplot',
                                                            fn7='transferSig_vs_FC_AUC',
                                                            fn8='transferSig_vs_FC_boxplot',
                                                            path = paste0("Results/2021/TB pre-vaccine/"))

# Save results
save(Iulio.2021.TB.transfer.sig.2021.results, 
     file="Results/2021/TB pre-vaccine/Iulio.2021.TB.transfer.sig.2021.results.RData")

```
