---
title: "Study 4 - Pre-vaccination inflammation and B-cell signalling predict age-related hyporesponse to hepatitis B vaccination (PMID: 26742691)"
author: "Mikkel Niklas Rasmussen"
date: "3/1/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Applying the BioAge, 15-gene signature and flow cytometric markers presented in study by Fourati e al. (PMID: 26742691) to the CMI-PB dataset

```{r load libaries, echo=FALSE}

setwd(dirname(rstudioapi::getSourceEditorContext()$path))

suppressMessages(library('data.table'))
suppressMessages(library('plyr'))
suppressMessages(library('dplyr'))
suppressMessages(library('tidyr'))
suppressMessages(library('Biobase'))
suppressMessages(library('DESeq2'))
suppressMessages(library('pROC'))
suppressMessages(library("ggplot2"))
suppressMessages(library("RCurl"))
suppressMessages(library("readr"))
suppressMessages(library("biomaRt"))
suppressMessages(library("ggpubr"))
suppressMessages(library("e1071"))
suppressMessages(library("readxl"))

# Load functions for evaluating gene signatures
source('../R/modelEvaluate.R')
```

## Load the CMI-PB datasets

```{r load data}

DATA_DIR_raw <- "../Data/raw"
DATA_DIR_full <- "../Data/processed/full"

# Info table (meta data)
meta.2020 <- fread(file.path(DATA_DIR_full, "meta.2020.pivoted.tsv"))
meta.2021 <- fread(file.path(DATA_DIR_full, "meta.2021.pivoted.tsv"))
specimen.table.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_specimen.csv"))
specimen.table.2021 <- fread(file.path(DATA_DIR_raw, "2021BD_specimen.csv"))

# Antibody titer data
ab.titer.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_ab_titer.csv"))
ab.titer.2021.day0 <- fread(file.path(DATA_DIR_raw, "2021BD_ab_titer.csv"))
ab.titer.2021.day14 <- fread(file.path(DATA_DIR_raw, "../2021_ab_titer_day14.csv"))

# Cell type CMI-PB info table
cell.types <- fread(file.path(DATA_DIR_raw, "../cell_types.txt"))
cell.types.2020 <- cell.types %>% 
   filter(dataset == "2020_dataset")
cell.types.2021 <- cell.types %>% 
   filter(dataset == "2021_dataset")

# Cell frequnecy data
cytof.2020 <- fread(file.path(DATA_DIR_raw, "2020LD_live_cell_percentages.csv"))
cytof.2021 <- fread(file.path(DATA_DIR_raw, "2021BD_live_cell_percentages.csv"))

# Gene expression data for day 0  - 2020
ge.TPM.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_tpm.2020.day0.pivoted.tsv"))
ge.RAW.2020.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2020.day0.pivoted.tsv"))

# Gene expression data for day 0  - 2021
ge.RAW.2021.day0 <- fread(file.path(DATA_DIR_full, "rnaseq_raw.2021.day0.pivoted.tsv"))


```

## Step 1: Preprocess CMI-PB 2020 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
abt.2020.all <- ab.titer.2020 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14, 30)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data day 0 and 14
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14)

# Subset antibody titer
abt.2020.subset <- abt.2020.all[abt.2020.all$subject_id %in% ids, ]
abt.2020.day0 <- abt.2020.subset[abt.2020.subset$timepoint == 0, ]
abt.2020.day14 <- abt.2020.subset[abt.2020.subset$timepoint == 14, ]

```

## Step 2: Preprocess CMI-PB 2020 cell population data

```{r}
#### Cell frequency data
# Merge cell frequnecy data with subject and day information
cytof.2020 <- cytof.2020 %>% 
   inner_join(., specimen.table.2020 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Select the subject that have data day 0 and 14
subjects.day0 <- unique(abt.2020.all[abt.2020.all$timepoint == 0, ]$subject_id)
subjects.day14 <- unique(abt.2020.all[abt.2020.all$timepoint == 14, ]$subject_id)
ids <- intersect(subjects.day0, subjects.day14)

# Join cell type and gating information with cell frequency data
cytof.2020.day0 <- cytof.2020[cytof.2020$subject_id %in% ids, ] %>% 
   inner_join(., cell.types.2020, by="cell_type_name") %>% 
   filter(timepoint == 0)

```

## Step 3: Transpose TPM normalized CMI-PB 2020 expression data and preprocess raw gene expression count data with DESeq2

```{r pre-process gene data}

# Tranpose the TPM normalized gene expression data
ge.TPM.2020.day0.T <- ge.TPM.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Tranpose the raw count gene expression data
ge.RAW.2020.day0.T <- ge.RAW.2020.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2020.day0 <- meta.2020[meta.2020$planned_day_relative_to_boost == 0, ]
meta.2020.day0.mask <- meta.2020.day0$subject_id %in% colnames(ge.RAW.2020.day0.T)
meta.2020.day0 <- meta.2020.day0[meta.2020.day0.mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2020.day0.T,
                              colData = meta.2020.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2020.day0 <- assay(vsd)

```

## Step 4: Evaluate the predictive power of the BioAge

```{r}

# Isotype-antigens which are considered the most important
selected.iso.ag <- c("IgG-PT", "IgG-FHA",   "IgG-PRN", "IgG1-PT",
                     "IgG1-FHA", "IgG4-PT", "IgG4-FHA")

################ TPM normalized gene expression ###############
# Calculate the BioAge score with TPM normalized gene expression
df.BioAge.2020.pred.TPM <- geneModuleCalc(gene.ex.data=ge.TPM.2020.day0.T,
                                          score='bioage',
                                          post.ab.data=abt.2020.day14,
                                          pre.ab.data=abt.2020.day0,
                                          filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with TPM normalized gene expression
Fourati.2016.BioAge.2020.results.TPM <- evaluateModel(df.prediction=df.BioAge.2020.pred.TPM,
                                                      N.perm = 10000, direction = "<",
                                                      add.to.title='- TPM normalized',
                                                      score.label = 'BioAge score',
                                                      fn1='BioAge_vs_day14_heatmap_TPM',
                                                      fn2='BioAge_vs_day14_correlation_TPM ',
                                                      fn3='BioAge_vs_FC_heatmap_TPM',
                                                      fn4='BioAge_vs_FC_correlation_TPM ',
                                                      fn5='BioAge_vs_day14_AUC_TPM',
                                                      fn6='BioAge_vs_day14_boxplot_TPM',
                                                      fn7='BioAge_vs_FC_AUC_TPM',
                                                      fn8='BioAge_vs_FC_boxplot_TPM',
                                                      path = 'Results/2020/TPM/BioAge/')

# Save results
save(Fourati.2016.BioAge.2020.results.TPM, 
     file = "Results/2020/TPM/BioAge/Fourati.2016.BioAge.2020.results.TPM.RData")

################ DESEQ normalized gene expression ###############
# Calculate the gene signature score with DESeq2 normalized gene expression
df.BioAge.2020.pred.DESEQ <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0, 
                                            score='bioage',
                                            post.ab.data=abt.2020.day14,
                                            pre.ab.data=abt.2020.day0,
                                            filter.ab.ag=selected.iso.ag)

# Evaluate gene signature with DESeq2 normalized gene expression
Fourati.2016.BioAge.2020.results.DESEQ <- evaluateModel(df.prediction=df.BioAge.2020.pred.DESEQ,
                                                        N.perm = 10000, direction = "missing",
                                                        add.to.title='- DESeq2 VST normalized',
                                                        score.label = 'BioAge score',
                                                        fn1='BioAge_vs_day14_heatmap_DESEQ',
                                                        fn2='BioAge_vs_day14_correlation_DESEQ',
                                                        fn3='BioAge_vs_FC_heatmap_DESEQ',
                                                        fn4='BioAge_vs_FC_correlation_DESEQ',
                                                        fn5='BioAge_vs_day14_AUC_DESEQ',
                                                        fn6='BioAge_vs_day14_boxplot_DESEQ',
                                                        fn7='BioAge_vs_FC_AUC_DESEQ',
                                                        fn8='BioAge_vs_FC_boxplot_DESEQ',
                                                        path='Results/2020/DESEQ/BioAge/')

# Save results
save(Fourati.2016.BioAge.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/BioAge/Fourati.2016.BioAge.2020.results.DESEQ.RData")
```

## Step 5: Evaluate the predictive power of the module M1 and M16 used in the BioAge 

```{r}

# Evaluate the predictive power of module M1
################ TPM normalized gene expression ###############
# Calculate the M1 BioAge score with TPM normalized gene expression
df.M1.2020.pred.TPM <- geneModuleCalc(gene.ex.data=ge.TPM.2020.day0.T,
                                      score='bioage',
                                      module=c(1), 
                                      post.ab.data=abt.2020.day14,
                                      pre.ab.data=abt.2020.day0,
                                      filter.ab.ag=selected.iso.ag)

# Evaluate M1 M1 score with TPM normalized gene expression
Fourati.2016.M1.2020.results.TPM <- evaluateModel(df.prediction=df.M1.2020.pred.TPM,
                                                  N.perm = 10000, direction = "missing",
                                                  add.to.title='- TPM normalized',
                                                  score.label = 'M1 BioAge score',
                                                  fn1='M1BioAge_vs_day14_heatmap_TPM',
                                                  fn2='M1BioAge_vs_day14_correlation_TPM ',
                                                  fn3='M1BioAge_vs_FC_heatmap_TPM',
                                                  fn4='M1BioAge_vs_FC_correlation_TPM ',
                                                  fn5='M1BioAge_vs_day14_AUC_TPM',
                                                  fn6='M1BioAge_vs_day14_boxplot_TPM',
                                                  fn7='M1BioAge_vs_FC_AUC_TPM',
                                                  fn8='M1BioAge_vs_FC_boxplot_TPM',
                                                  path = 'Results/2020/TPM/M1/')

# Save results
save(Fourati.2016.M1.2020.results.TPM, 
     file = "Results/2020/TPM/M1/Fourati.2016.M1.2020.results.TPM.RData")

# Evaluate the predictive power of module M1 and M16 together
# Calculate the M1+M16 BioAge score with TPM normalized gene expression
df.M1.M16.2020.pred.TPM <- geneModuleCalc(gene.ex.data=ge.TPM.2020.day0.T,
                                     score='bioage',
                                     module=c(1, 16), 
                                     post.ab.data=abt.2020.day14,
                                     pre.ab.data=abt.2020.day0,
                                     filter.ab.ag=selected.iso.ag)

# Evaluate M1+M16 BioAge score with TPM normalized gene expression
Fourati.2016.M1.M16.2020.results.TPM <- evaluateModel(df.prediction=df.M1.M16.2020.pred.TPM,
                                                      N.perm = 10000, direction = "missing",
                                                      add.to.title='- TPM normalized',
                                                      score.label = 'M1+M16 BioAge score',
                                                      fn1='M1M16BioAge_vs_day14_heatmap_TPM',
                                                      fn2='M1M16BioAge_vs_day14_correlation_TPM ',
                                                      fn3='M1M16BioAge_vs_FC_heatmap_TPM',
                                                      fn4='M1M16BioAge_vs_FC_correlation_TPM ',
                                                      fn5='M1M16BioAge_vs_day14_AUC_TPM',
                                                      fn6='M1M16BioAge_vs_day14_boxplot_TPM',
                                                      fn7='M1M16BioAge_vs_FC_AUC_TPM',
                                                      fn8='M1M16BioAge_vs_FC_boxplot_TPM',
                                                      path = 'Results/2020/TPM/M1+M16/')

# Save results
save(Fourati.2016.M1.M16.2020.results.TPM, 
     file = "Results/2020/TPM/M1+M16/Fourati.2016.M1.M16.2020.results.TPM.RData")

################ DESEQ normalized gene expression ###############
# Calculate the M1 BioAge score with DESeq2 normalized gene expression
df.M1.2020.pred.DESEQ <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0,
                                        score='bioage',
                                        module=c(1), 
                                        post.ab.data=abt.2020.day14,
                                        pre.ab.data=abt.2020.day0,
                                        filter.ab.ag=selected.iso.ag)

# Evaluate M1 BioAge score with DESeq2 normalized gene expression
Fourati.2016.M1.2020.results.DESEQ <- evaluateModel(df.prediction=df.M1.2020.pred.DESEQ,
                                                    N.perm = 10000, direction = "missing",
                                                    add.to.title='- DESeq2 VST normalized',
                                                    score.label = 'M1 BioAge score',
                                                    fn1='M1BioAge_vs_day14_heatmap_DESEQ',
                                                    fn2='M1BioAge_vs_day14_correlation_DESEQ ',
                                                    fn3='M1BioAge_vs_FC_heatmap_DESEQ',
                                                    fn4='M1BioAge_vs_FC_correlation_DESEQ ',
                                                    fn5='M1BioAge_vs_day14_AUC_DESEQ',
                                                    fn6='M1BioAge_vs_day14_boxplot_DESEQ',
                                                    fn7='M1BioAge_vs_FC_AUC_DESEQ',
                                                    fn8='M1BioAge_vs_FC_boxplot_DESEQ',
                                                    path = 'Results/2020/DESEQ/M1/')

# Save results
save(Fourati.2016.M1.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/M1/Fourati.2016.M1.2020.results.DESEQ.RData")

# Evaluate the predictive power of module M1 and M16 together
# Calculate the M1+M16 BioAge score with DESeq2 normalized gene expression
df.M1.M16.2020.pred.DESEQ <- geneModuleCalc(gene.ex.data=ge.norm.2020.day0,
                                            score='bioage',
                                            module=c(1, 16), 
                                            post.ab.data=abt.2020.day14,
                                            pre.ab.data=abt.2020.day0,
                                            filter.ab.ag=selected.iso.ag)

# Evaluate M1+M16 BioAge score with DESeq2 normalized gene expression
Fourati.2016.M1.M16.2020.results.DESEQ <- evaluateModel(df.prediction=df.M1.M16.2020.pred.DESEQ,
                                                        N.perm = 10000, direction = "missing",
                                                        add.to.title='- DESeq2 VST normalized',
                                                        score.label = 'M1+M16 BioAge score',
                                                        fn1='M1M16BioAge_vs_day14_heatmap_DESEQ',
                                                        fn2='M1M16BioAge_vs_day14_correlation_DESEQ ',
                                                        fn3='M1M16BioAge_vs_FC_heatmap_DESEQ',
                                                        fn4='M1M16BioAge_vs_FC_correlation_DESEQ ',
                                                        fn5='M1M16BioAge_vs_day14_AUC_DESEQ',
                                                        fn6='M1M16BioAge_vs_day14_boxplot_DESEQ',
                                                        fn7='M1M16BioAge_vs_FC_AUC_DESEQ',
                                                        fn8='M1M16BioAge_vs_FC_boxplot_DESEQ',
                                                        path = 'Results/2020/DESEQ/M1+M16/')

# Save results
save(Fourati.2016.M1.M16.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/M1+M16/Fourati.2016.M1.M16.2020.results.DESEQ.RData")

```

## Step 6: Evaluate the naive bayes classifier with the 15-gene signature as input

```{r}

# Load Naive Bayes classifier
load(file='em131.naiveBayes.RData')
NB.fit <- fit

################ TPM normalized gene expression ###############
# Calculate the gene signature score with DESeq2 normalized gene expression
df.NB.2020.pred.TPM <- mlModelPredict(gene.ex.data=ge.TPM.2020.day0.T,
                                      model=NB.fit, study='fourati_2016_NB',
                                      post.ab.data=abt.2020.day14, 
                                      pre.ab.data=abt.2020.day0,
                                      filter.ab.ag=selected.iso.ag, 
                                      verbose=TRUE)

# Evaluate gene signature with TPM normalized gene expression
Fourati.2016.NB.2020.results.TPM <- evaluateModel(df.prediction=df.NB.2020.pred.TPM,
                                                  N.perm = 10000, direction = "missing",
                                                  add.to.title='- TPM normalized',
                                                  score.label = 'Ratio between post-probalities',
                                                  fn1='NBpred_vs_day14_heatmap',
                                                  fn2='NBpred_vs_day14_correlation',
                                                  fn3='NBpred_vs_FC_heatmap',
                                                  fn4='NBpred_vs_FC_correlation',
                                                  fn5='NBpred_vs_day14_AUC',
                                                  fn6='NBpred_vs_day14_boxplot',
                                                  fn7='NBpred_vs_FC_AUC',
                                                  fn8='NBpred_vs_FC_boxplot',
                                                  path='Results/2020/TPM/Naive_bayes/')

# Save results
save(Fourati.2016.NB.2020.results.TPM, 
     file = "Results/2020/TPM/Naive_bayes/Fourati.2016.NB.2020.results.TPM.RData")

################ DESEQ normalized gene expression ###############
# Calculate the gene signature score with DESeq2 normalized gene expression
df.NB.2020.pred.DESEQ <- mlModelPredict(gene.ex.data=ge.norm.2020.day0,
                                        model=NB.fit, study='fourati_2016_NB',
                                        post.ab.data=abt.2020.day14, 
                                        pre.ab.data=abt.2020.day0,
                                        filter.ab.ag=selected.iso.ag, 
                                        verbose=TRUE)

# Evaluate gene signature with DESeq2 normalized gene expression
Fourati.2016.NB.2020.results.DESEQ <- evaluateModel(df.prediction=df.NB.2020.pred.DESEQ,
                                                    N.perm = 10000, direction = "missing",
                                                    add.to.title='- DESeq2 VST normalized',
                                                    score.label = 'Ratio between post-probalities',
                                                    fn1='NBpred_vs_day14_heatmap',
                                                    fn2='NBpred_vs_day14_correlation',
                                                    fn3='NBpred_vs_FC_heatmap',
                                                    fn4='NBpred_vs_FC_correlation',
                                                    fn5='NBpred_vs_day14_AUC',
                                                    fn6='NBpred_vs_day14_boxplot',
                                                    fn7='NBpred_vs_FC_AUC',
                                                    fn8='NBpred_vs_FC_boxplot',
                                                    path='Results/2020/DESEQ/Naive_bayes/')

# Save results
save(Fourati.2016.NB.2020.results.DESEQ, 
     file = "Results/2020/DESEQ/Naive_bayes/Fourati.2016.NB.2020.results.DESEQ.RData")

```

## Step 7: Evaluate the multivariate regression model using the 4 flow cytometric markers as input

```{r}

# Manually select the 4 cell populations in the CMI-PB closest ones reported in 
# the orignial manuscript
 
# "B cells.PCT IgG+ memory B cells"   
print("Original study - gating definition:")
print("CD3-CD19+HLADR+CD27+CD10-CD20+IgG+")
print("Closest cell population in CMI-PB 2020 dataset - gatinge definition:")
cell.types.2020[cell.types.2020$cell_type_name == "Bcells", ]$gating_definition

# "B cells.PCT switched IgG+ MK memory B cells"
print("Original study - gating definition:")
print("CD3-CD19+HLADR+CD27+IgG+")
print("Closest cell population in CMI-PB 2020 dataset - gatinge definition:")
cell.types.2020[cell.types.2020$cell_type_name == "ASCs (Plasmablasts)", ]$gating_definition

# "T cells.PCT TEM2 in CD4"   
print("Original study - gating definition:")
print("CD3+CD4+CD45RA-CD27-CCR7-CD28-")
print("Closest cell population in CMI-PB 2020 dataset - gatinge definition:")
cell.types.2020[cell.types.2020$cell_type_name == 'TemCD4', ]$gating_definition

# "Innate cells.MdFI CD40 in PDC"
print("Original study - gating definition:")
print("CD40")
print("Closest cell population in CMI-PB 2020 dataset - gatinge definition:")
cell.types.2020[cell.types.2020$cell_type_name == 'pDC', ]$gating_definition

cell.populations.2020 <- c("Bcells", 'ASCs (Plasmablasts)', "TemCD4", "pDC")

# Load logistic regression model
load(file='em131.logisticReg.RData')
LR.fit <- fit

# Calculate the output of the logistic regression model with 4 cell populations
df.LR.2020.pred <- mlModelPredict(cytof.data=cytof.2020.day0,
                                  populations=cell.populations.2020,
                                  study='fourati_2016_LR', 
                                  model=LR.fit, year="2020",
                                  filter.ab.ag=selected.iso.ag,
                                  post.ab.data=abt.2020.day14,
                                  pre.ab.data=abt.2020.day0)

# Evaluate the logistic regression model with 4 cell populations
Fourati.2016.LR.2020.results <- evaluateModel(df.prediction=df.LR.2020.pred, 
                                              direction = 'missing', N.perm = 10000,
                                              score.label='LR output - scale of the linear predictors',
                                              add.to.title = " - Logistic regression using cell populations",
                                              fn1='LRpred_vs_day14_heatmap',
                                              fn2='LRpred_vs_day14_correlation',
                                              fn3='LRpred_vs_FC_heatmap',
                                              fn4='LRpred_vs_FC_correlation',
                                              fn5='LRpred_vs_day14_AUC',
                                              fn6='LRpred_vs_day14_boxplot',
                                              fn7='LRpred_vs_FC_AUC',
                                              fn8='LRpred_vs_FC_boxplot',
                                              path = 'Results/2020/LR/')

# Save results
save(Fourati.2016.LR.2020.results, 
     file = "Results/2020/LR/Fourati.2016.LR.2020.results.RData")

```

## Step 8: Pre-process CMI-PB 2021 antibody titer data

```{r}

# Normalized antibody titer data by baseline median
ab.titer.day0.2021.all <- ab.titer.2021.day0 %>% 
   filter(antigen %in% c('PT', 'PRN', 'FHA', 'FIM2/3')) %>% 
   unite(isotype_antigen, c("isotype", "antigen"), sep = "-") %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

ab.titer.subject.ids <- unique(ab.titer.day0.2021.all$subject_id)
abt.2021.day0 <- ab.titer.day0.2021.all[ab.titer.day0.2021.all$timepoint == 0, ]
abt.2021.day0 <- abt.2021.day0 %>% 
   filter(isotype_antigen %in% selected.iso.ag)

# Change format of the day 14 antibody titers
abt.2021.day14 <- ab.titer.2021.day14 %>% 
   separate(task, into=c("task", "antigen"), sep = "\\) ") %>% 
   dplyr::mutate(isotype_antigen = recode(antigen, 
                                          `Pertussis Toxin` = "IgG-PT", 
                                          `FHA` = 'IgG-FHA', 
                                          `Pertactin` =  'IgG-PRN',
                                          `IgG1 - Pertussis toxin` = "IgG1-PT",
                                          `IgG1 - FHA` = "IgG1-FHA",
                                          `IgG4 - Pertussis toxin` = "IgG4-PT",
                                          `IgG4 - FHA` = "IgG4-FHA"))

```

## Step 9: Preprocess raw CMI-PB 2021 gene expression count data with DESeq2

```{r pre-process gene data}

# Subset gene expression data based on subjects with antibody titers
ge.RAW.2021.day0 <- ge.RAW.2021.day0[ge.RAW.2021.day0$subject_id 
                                     %in% ab.titer.subject.ids, ]

# Tranpose the raw cout gene expression data
ge.RAW.2021.day0.T <- ge.RAW.2021.day0 %>% 
   tibble::column_to_rownames('subject_id') %>% 
   t()

# Subset antibody titers data
meta.2021.day0 <- meta.2021[meta.2021$planned_day_relative_to_boost == 0, ]
meta.2021.day0..mask <- meta.2021.day0$subject_id %in% colnames(ge.RAW.2021.day0.T)
meta.2021.day0 <- meta.2021.day0[meta.2021.day0..mask, ]

# Create DESeq dataset object
dds <- DESeqDataSetFromMatrix(countData = ge.RAW.2021.day0.T,
                              colData = meta.2021.day0,
                              design = ~ 1)

# DESeq2 normalization with variance stabilizing transformation (vst)
cds <- estimateSizeFactors(dds) ## estimate size factor
cdsBlind <- estimateDispersions(cds)
vsd <- vst(cdsBlind, blind=TRUE)
ge.norm.2021.day0 <- assay(vsd)

```

## Step 10: Preprocess CMI-PB 2021 cell population data

```{r}
#### Cell frequency data
# Merge cell frequnecy data with subject and day information
cytof.2021 <- cytof.2021 %>% 
   inner_join(., specimen.table.2021 %>% dplyr::select(specimen_id, subject_id, 
                                                  planned_day_relative_to_boost), 
              by='specimen_id') %>% 
   filter(planned_day_relative_to_boost %in% c(0, 1, 3, 7, 14)) %>% 
   dplyr::rename(timepoint = planned_day_relative_to_boost) %>% 
   tibble()

# Subset cell frequency data based on subjects with antibody titers
cytof.2021 <- cytof.2021[cytof.2021$subject_id %in% ab.titer.subject.ids, ]

# Join cell type and gating information with cell frequency data
cytof.2021.day0 <- cytof.2021 %>% 
   inner_join(., cell.types.2021, by="cell_type_name") %>% 
   filter(timepoint == 0)


# Manually select the 4 cell populations in the CMI-PB 2021 data closest ones 
# reported in the orignial manuscript
 
# "B cells.PCT IgG+ memory B cells"   
print("Original study - gating definition:")
print("CD3-CD19+HLADR+CD27+CD10-CD20+IgG+")
print("Closest cell population in CMI-PB 2021 dataset - gatinge definition:")
cell.types.2021[cell.types.2021$cell_type_name == 'Bcells', ]$gating_definition

# "B cells.PCT switched IgG+ MK memory B cells"
print("Original study - gating definition:")
print("CD3-CD19+HLADR+CD27+IgG+")
print("Closest cell population in CMI-PB 2021 dataset - gatinge definition:")
cell.types.2021[cell.types.2021$cell_type_name == 'Memory B cells', ]$gating_definition

# "T cells.PCT TEM2 in CD4"   
print("Original study - gating definition:")
print("CD3+CD4+CD45RA-CD27-CCR7-CD28-")
print("Closest cell population in CMI-PB 2021 dataset - gatinge definition:")
cell.types.2021[cell.types.2021$cell_type_name == 'TemCD4', ]$gating_definition

# "Innate cells.MdFI CD40 in PDC"
print("Original study - gating definition:")
print("CD40")
print("Closest cell population in CMI-PB 2021 dataset - gatinge definition:")
cell.types.2021[cell.types.2021$cell_type_name == 'pDC', ]$gating_definition

cell.populations.2021 <- c("Bcells", "Memory B cells", "TemCD4", "pDC")

```

## Step 11: Make prediction for CMI-PB 2021 dataset

```{r}

# Calculate the BioAge score with DESeq2 normalized gene expression
df.BioAge.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                      score='bioage')


# Calculate the M1 BioAge score with DESeq2 normalized gene expression
df.M1.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0,
                                 score='bioage',
                                 module=c(1))


# Calculate the M1+M16 BioAge score with DESeq2 normalized gene expression
df.M1.M16.pred.2021 <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0,
                                       score='bioage',
                                       module=c(1, 16))

# Calculate the gene signature score with DESeq2 normalized gene expression
df.NB.pred.2021 <- mlModelPredict(gene.ex.data=ge.norm.2021.day0,
                                   model=NB.fit, study='fourati_2016_NB',
                                   verbose=TRUE)

# Calculate the output of the logistic regression model with 4 cell populations
df.LR.pred.2021 <- mlModelPredict(cytof.data=cytof.2021.day0,
                                  populations=cell.populations.2021,
                                  study='fourati_2016_LR', 
                                  model=LR.fit,
                                  year="2021")


# Read submisson format
submission.format <- fread("../Data/2021_submission_format.tsv") %>% 
   dplyr::rename(subject_id = Subject_ID)

# BioAge predictions
Fourati.2016.BioAge.2021.pred <- submission.format %>% 
   left_join(., df.BioAge.pred.2021, by="subject_id")
Fourati.2016.BioAge.2021.pred[, c(5:11)] <- Fourati.2016.BioAge.2021.pred$rank.predictor
Fourati.2016.BioAge.2021.pred <- Fourati.2016.BioAge.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Fourati.2016.BioAge.2021.pred, 
            file = "Results/2021_predictions/fourati_2016_BioAge_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# M1 predictions
Fourati.2016.M1.2021.pred <- submission.format %>% 
   left_join(., df.M1.pred.2021, by="subject_id")
Fourati.2016.M1.2021.pred[, c(5:11)] <- Fourati.2016.M1.2021.pred$rank.predictor
Fourati.2016.M1.2021.pred <- Fourati.2016.M1.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Fourati.2016.M1.2021.pred, 
            file = "Results/2021_predictions/fourati_2016_M1_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# M1+M16 predictions
Fourati.2016.M1.M16.2021.pred <- submission.format %>% 
   left_join(., df.M1.M16.pred.2021, by="subject_id")
Fourati.2016.M1.M16.2021.pred[, c(5:11)] <- Fourati.2016.M1.M16.2021.pred$rank.predictor
Fourati.2016.M1.M16.2021.pred <- Fourati.2016.M1.M16.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Fourati.2016.M1.M16.2021.pred, 
            file = "Results/2021_predictions/fourati_2016_M1+M16_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)


# Naive Bayes predictions
Fourati.2016.NB.2021.pred <- submission.format %>% 
   left_join(., df.NB.pred.2021, by="subject_id")
Fourati.2016.NB.2021.pred[, c(5:11)] <- Fourati.2016.NB.2021.pred$rank.predictor
Fourati.2016.NB.2021.pred <- Fourati.2016.NB.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Fourati.2016.NB.2021.pred, 
            file = "Results/2021_predictions/fourati_2016_NB_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)

# Logistic regression predictions
Fourati.2016.LR.2021.pred <- submission.format %>% 
   left_join(., df.LR.pred.2021, by="subject_id")
Fourati.2016.LR.2021.pred[, c(5:11)] <- Fourati.2016.LR.2021.pred$rank.predictor
Fourati.2016.LR.2021.pred <- Fourati.2016.LR.2021.pred[,c(1:18)] %>% 
   dplyr::rename(Subject_ID = subject_id)
# Save predictions
write.table(Fourati.2016.LR.2021.pred, 
            file = "Results/2021_predictions/fourati_2016_LR_CMI-PB_2021.tsv",
            quote=FALSE, sep='\t', row.names = FALSE)


```

## Step 12: Evaluate the performance of the BioAge (M1 / M1+M16), Naive Bayes, and Logistic regression predictions for the CMI-PB 2021 data

```{r}

# Calculate the BioAge score with DESeq2 normalized gene expression
df.BioAge.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0, 
                                      score='bioage',
                                      post.ab.data=abt.2021.day14,
                                      pre.ab.data=abt.2021.day0,
                                      filter.ab.ag=selected.iso.ag)


# Evaluate gene signature with DESeq2 normalized gene expression
Fourati.2016.BioAge.2021.results <- evaluateModel(df.prediction=df.BioAge.2021.pred,
                                                  N.perm = 10000, direction = "missing",
                                                  add.to.title='- DESeq2 VST normalized',
                                                  score.label = 'BioAge score',
                                                  fn1='BioAge_vs_day14_heatmap',
                                                  fn2='BioAge_vs_day14_correlation',
                                                  fn3='BioAge_vs_FC_heatmap',
                                                  fn4='BioAge_vs_FC_correlation',
                                                  fn5='BioAge_vs_day14_AUC',
                                                  fn6='BioAge_vs_day14_boxplot',
                                                  fn7='BioAge_vs_FC_AUC',
                                                  fn8='BioAge_vs_FC_boxplot',
                                                  path='Results/2021/BioAge/')

# Save results
save(Fourati.2016.BioAge.2021.results, 
     file = "Results/2021/BioAge/Fourati.2016.BioAge.2021.results.RData")


# Calculate the M1 BioAge score with DESeq2 normalized gene expression
df.M1.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0,
                                  score='bioage',
                                  module=c(1),
                                  post.ab.data=abt.2021.day14,
                                  pre.ab.data=abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag)

# Evaluate M1 BioAge score with DESeq2 normalized gene expression
Fourati.2016.M1.2021.results <- evaluateModel(df.prediction=df.M1.2021.pred,
                                              N.perm = 10000, direction = "missing",
                                              add.to.title='- DESeq2 VST normalized',
                                              score.label = 'M1 BioAge score',
                                              fn1='M1BioAge_vs_day14_heatmap',
                                              fn2='M1BioAge_vs_day14_correlation',
                                              fn3='M1BioAge_vs_FC_heatmap',
                                              fn4='M1BioAge_vs_FC_correlation',
                                              fn5='M1BioAge_vs_day14_AUC',
                                              fn6='M1BioAge_vs_day14_boxplot',
                                              fn7='M1BioAge_vs_FC_AUC',
                                              fn8='M1BioAge_vs_FC_boxplot',
                                              path = 'Results/2021/M1/')
# Save results
save(Fourati.2016.M1.2021.results, 
     file = "Results/2021/M1/Fourati.2016.M1.2021.results.RData")


# Calculate the M1+M16 BioAge score with DESeq2 normalized gene expression
df.M1.M16.2021.pred <- geneModuleCalc(gene.ex.data=ge.norm.2021.day0,
                                      score='bioage',
                                      module=c(1, 16),
                                      post.ab.data=abt.2021.day14,
                                      pre.ab.data=abt.2021.day0,
                                      filter.ab.ag=selected.iso.ag)

# Evaluate M1+M16 BioAge score with DESeq2 normalized gene expression
Fourati.2016.M1.M16.2021.results <- evaluateModel(df.prediction=df.M1.M16.2021.pred,
                                                  N.perm = 10000, direction = "missing",
                                                  add.to.title='- DESeq2 VST normalized',
                                                  score.label = 'M1+M16 BioAge score',
                                                  fn1='M1M16BioAge_vs_day14_heatmap',
                                                  fn2='M1M16BioAge_vs_day14_correlation',
                                                  fn3='M1M16BioAge_vs_FC_heatmap',
                                                  fn4='M1M16BioAge_vs_FC_correlation',
                                                  fn5='M1M16BioAge_vs_day14_AUC',
                                                  fn6='M1M16BioAge_vs_day14_boxplot',
                                                  fn7='M1M16BioAge_vs_FC_AUC',
                                                  fn8='M1M16BioAge_vs_FC_boxplot',
                                                  path = 'Results/2021/M1+M16/')

# Save results
save(Fourati.2016.M1.M16.2021.results, 
     file = "Results/2021/M1+M16/Fourati.2016.M1.M16.2021.results.RData")

# Calculate the gene signature score with DESeq2 normalized gene expression
df.NB.2021.pred <- mlModelPredict(gene.ex.data=ge.norm.2021.day0,
                                  model=NB.fit, 
                                  study='fourati_2016_NB',
                                  post.ab.data=abt.2021.day14,
                                  pre.ab.data=abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag,
                                  verbose=TRUE)

# Evaluate gene signature with DESeq2 normalized gene expression
Fourati.2016.NB.2021.results <- evaluateModel(df.prediction=df.NB.2021.pred,
                                              N.perm = 10000, direction = "missing",
                                              add.to.title='- DESeq2 VST normalized',
                                              score.label = 'Ratio between post-probalities',
                                              fn1='NBpred_vs_day14_heatmap',
                                              fn2='NBpred_vs_day14_correlation',
                                              fn3='NBpred_vs_FC_heatmap',
                                              fn4='NBpred_vs_FC_correlation',
                                              fn5='NBpred_vs_day14_AUC',
                                              fn6='NBpred_vs_day14_boxplot',
                                              fn7='NBpred_vs_FC_AUC',
                                              fn8='NBpred_vs_FC_boxplot',
                                              path='Results/2021/Naive_bayes/')

# Save results
save(Fourati.2016.NB.2021.results, 
     file = "Results/2021/Naive_bayes/Fourati.2016.NB.2021.results.RData")

# Calculate the output of the logistic regression model with 4 cell populations
df.LR.2021.pred <- mlModelPredict(cytof.data=cytof.2021.day0,
                                  populations=cell.populations.2021,
                                  study='fourati_2016_LR', 
                                  model=LR.fit,
                                  year="2021",
                                  post.ab.data=abt.2021.day14,
                                  pre.ab.data=abt.2021.day0,
                                  filter.ab.ag=selected.iso.ag,
                                  verbose=TRUE)

# Evaluate the logistic regression model with 4 cell populations
Fourati.2016.LR.2021.results <- evaluateModel(df.prediction=df.LR.2021.pred, 
                                              direction = 'missing', N.perm = 10000,
                                              score.label='LR output - scale of the linear predictors',
                                              add.to.title = " - Logistic regression using cell populations",
                                              fn1='LRpred_vs_day14_heatmap',
                                              fn2='LRpred_vs_day14_correlation',
                                              fn3='LRpred_vs_FC_heatmap',
                                              fn4='LRpred_vs_FC_correlation',
                                              fn5='LRpred_vs_day14_AUC',
                                              fn6='LRpred_vs_day14_boxplot',
                                              fn7='LRpred_vs_FC_AUC',
                                              fn8='LRpred_vs_FC_boxplot',
                                              path = 'Results/2021/LR/')

# Save results
save(Fourati.2016.LR.2021.results, 
     file = "Results/2021/LR/Fourati.2016.LR.2021.results.RData")


```

